// @a5c-ai/babysitter-sdk/dist/runtime/hooks/runtime.js
"use strict";
/**
 * Runtime hook integration helpers
 *
 * Provides safe hook calling with error handling for SDK runtime.
 * Hook failures are logged but do not break orchestration.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.callRuntimeHook = callRuntimeHook;
exports.createRuntimeHookPayload = createRuntimeHookPayload;
const dispatcher_1 = require("../../hooks/dispatcher");
/**
 * Safely call a hook from SDK runtime with error handling.
 *
 * Hook failures are logged but do not throw - orchestration continues.
 *
 * @param hookType - The type of hook to call
 * @param payload - The hook payload
 * @param options - Runtime options (cwd, timeout, logger)
 * @returns HookResult with execution details
 */
async function callRuntimeHook(hookType, payload, options) {
    const { cwd, timeout = 30000, logger } = options;
    try {
        // Ensure timestamp is present
        const fullPayload = {
            ...payload,
            hookType,
            timestamp: payload.timestamp || new Date().toISOString(),
        };
        const result = await (0, dispatcher_1.callHook)({
            hookType,
            payload: fullPayload,
            cwd,
            timeout,
        });
        // Log hook execution if logger provided
        if (logger && result.executedHooks.length > 0) {
            logger(`[hooks] Executed ${result.executedHooks.length} hook(s) for ${hookType}`);
        }
        return result;
    }
    catch (error) {
        // Hook failures should not break orchestration
        const errorMessage = error instanceof Error ? error.message : String(error);
        if (logger) {
            logger(`[hooks] Hook execution failed for ${hookType}: ${errorMessage}`);
        }
        // Return a failure result instead of throwing
        return {
            hookType,
            success: false,
            error: errorMessage,
            executedHooks: [],
        };
    }
}
/**
 * Create a hook payload with automatic timestamp.
 */
function createRuntimeHookPayload(hookType, data) {
    return {
        ...data,
        hookType,
        timestamp: new Date().toISOString(),
    };
}
