// @a5c-ai/babysitter-sdk/dist/runtime/createRun.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createRun = createRun;
const path_1 = __importDefault(require("path"));
const crypto_1 = __importDefault(require("crypto"));
const createRunDir_1 = require("../storage/createRunDir");
const journal_1 = require("../storage/journal");
const lock_1 = require("../storage/lock");
const paths_1 = require("../storage/paths");
const ulids_1 = require("../storage/ulids");
const runtime_1 = require("./hooks/runtime");
async function createRun(options) {
    const runId = options.runId ?? (0, ulids_1.nextUlid)();
    validateRunId(runId);
    const runDir = (0, paths_1.getRunDir)(options.runsDir, runId);
    const normalizedEntrypoint = normalizeEntrypoint(runDir, options.process.importPath, options.process.exportName);
    const requestId = options.request ?? options.process.processId ?? runId;
    const providedProof = typeof options.metadata?.completionProof === "string" ? options.metadata.completionProof : undefined;
    const completionProof = providedProof ?? crypto_1.default.randomBytes(16).toString("hex");
    const extraMetadata = {
        ...options.metadata,
        completionProof,
    };
    const { metadata } = await (0, createRunDir_1.createRunDir)({
        runsRoot: options.runsDir,
        runId,
        request: requestId,
        processId: options.process.processId,
        processRevision: options.processRevision,
        layoutVersion: options.layoutVersion,
        inputs: options.inputs,
        entrypoint: normalizedEntrypoint,
        processPath: normalizedEntrypoint.importPath,
        extraMetadata,
        prompt: options.prompt,
    });
    let lockAcquired = false;
    try {
        await (0, lock_1.acquireRunLock)(runDir, options.lockOwner ?? "runtime:createRun");
        lockAcquired = true;
        const eventPayload = {
            runId,
            processId: metadata.processId,
            entrypoint: metadata.entrypoint,
        };
        if (metadata.processRevision) {
            eventPayload.processRevision = metadata.processRevision;
        }
        if (options.inputs !== undefined) {
            eventPayload.inputsRef = paths_1.INPUTS_FILE;
        }
        if (options.prompt !== undefined) {
            eventPayload.prompt = options.prompt;
        }
        await (0, journal_1.appendEvent)({
            runDir,
            eventType: "RUN_CREATED",
            event: eventPayload,
        });
    }
    finally {
        if (lockAcquired) {
            await (0, lock_1.releaseRunLock)(runDir);
        }
    }
    // Call on-run-start hook
    const entryString = metadata.entrypoint.exportName
        ? `${metadata.entrypoint.importPath}#${metadata.entrypoint.exportName}`
        : metadata.entrypoint.importPath;
    // Call hook from project root (parent of .a5c dir) where plugins/ is located
    // runDir is like: /path/to/project/.a5c/runs/<runId>
    // So we need 3 levels up: runs -> .a5c -> project
    const projectRoot = path_1.default.dirname(path_1.default.dirname(path_1.default.dirname(runDir)));
    await (0, runtime_1.callRuntimeHook)("on-run-start", {
        runId,
        processId: metadata.processId,
        entry: entryString,
        inputs: options.inputs,
    }, {
        cwd: projectRoot,
        logger: options.logger,
    });
    return {
        runId,
        runDir,
        metadata,
    };
}
function validateRunId(runId) {
    if (typeof runId !== "string" || runId.trim() === "") {
        throw new Error("runId must be a non-empty string");
    }
}
function normalizeEntrypoint(runDir, importPath, exportName) {
    const entryImport = toRunRelativePosix(runDir, importPath);
    return {
        importPath: entryImport,
        exportName: exportName ?? "process",
    };
}
function toRunRelativePosix(runDir, importPath) {
    const relative = path_1.default.isAbsolute(importPath) ? path_1.default.relative(runDir, importPath) : importPath;
    if (!relative || relative === ".") {
        throw new Error("Entrypoint import path must reference a file");
    }
    return path_1.default.posix.normalize(relative.replace(/\\/g, "/"));
}
