// @a5c-ai/babysitter-sdk/dist/runtime/errorUtils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.serializeUnknownError = serializeUnknownError;
exports.toSerializedEffectError = toSerializedEffectError;
function serializeUnknownError(error) {
    if (error instanceof Error) {
        return {
            name: error.name ?? "Error",
            message: error.message ?? "Unknown error",
            stack: error.stack,
        };
    }
    if (typeof error === "object" && error) {
        return {
            name: error.constructor?.name ?? "Error",
            message: JSON.stringify(error),
            data: error,
        };
    }
    return {
        name: "Error",
        message: String(error),
    };
}
function toSerializedEffectError(error) {
    if (error && typeof error === "object" && "name" in error) {
        const err = error;
        return {
            name: err.name ?? "Error",
            message: err.message ?? "Task failed",
            stack: typeof err.stack === "string" ? err.stack : undefined,
            data: err.data,
        };
    }
    const serialized = serializeUnknownError(error);
    return {
        name: serialized.name,
        message: serialized.message,
        stack: serialized.stack,
        data: serialized.data,
    };
}
