// @a5c-ai/babysitter-sdk/dist/runtime/exceptions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InvalidSleepTargetError = exports.InvalidTaskDefinitionError = exports.MissingProcessContextError = exports.RunFailedError = exports.InvocationCollisionError = exports.ParallelPendingError = exports.EffectPendingError = exports.EffectRequestedError = exports.BabysitterIntrinsicError = exports.BabysitterRuntimeError = exports.ERROR_TEMPLATES = exports.ERROR_CATEGORY_DESCRIPTIONS = exports.ErrorCategory = void 0;
exports.interpolateTemplate = interpolateTemplate;
exports.createErrorMessage = createErrorMessage;
exports.suggestCommand = suggestCommand;
exports.suggestFlag = suggestFlag;
exports.suggestFix = suggestFix;
exports.formatErrorWithContext = formatErrorWithContext;
exports.formatNextSteps = formatNextSteps;
exports.toStructuredError = toStructuredError;
exports.isIntrinsicError = isIntrinsicError;
exports.isBabysitterError = isBabysitterError;
exports.rehydrateSerializedError = rehydrateSerializedError;
const batching_1 = require("../tasks/batching");
// ============================================================================
// Error Categories
// ============================================================================
/**
 * Error category enum for classifying errors by their source and type.
 * This helps users understand the nature of errors and how to address them.
 */
var ErrorCategory;
(function (ErrorCategory) {
    /** Configuration errors - issues with settings, environment, or setup */
    ErrorCategory["Configuration"] = "CONFIGURATION";
    /** Validation errors - input data or parameter validation failures */
    ErrorCategory["Validation"] = "VALIDATION";
    /** Runtime errors - errors during execution of processes or tasks */
    ErrorCategory["Runtime"] = "RUNTIME";
    /** External errors - failures from external services, APIs, or dependencies */
    ErrorCategory["External"] = "EXTERNAL";
    /** Internal errors - unexpected internal failures, bugs, or invariant violations */
    ErrorCategory["Internal"] = "INTERNAL";
})(ErrorCategory || (exports.ErrorCategory = ErrorCategory = {}));
/**
 * Human-readable descriptions for each error category
 */
exports.ERROR_CATEGORY_DESCRIPTIONS = {
    [ErrorCategory.Configuration]: "Configuration or setup issue",
    [ErrorCategory.Validation]: "Input validation failure",
    [ErrorCategory.Runtime]: "Runtime execution error",
    [ErrorCategory.External]: "External service or dependency failure",
    [ErrorCategory.Internal]: "Internal error (please report as a bug)",
};
/**
 * Built-in error templates for common error scenarios
 */
exports.ERROR_TEMPLATES = {
    MISSING_REQUIRED_FLAG: {
        pattern: "Missing required flag: {{flag}}",
        category: ErrorCategory.Validation,
        defaultNextSteps: ["Add the missing flag to your command", "Run with --help to see all available options"],
    },
    INVALID_FLAG_VALUE: {
        pattern: "Invalid value for {{flag}}: {{value}}. Expected {{expected}}",
        category: ErrorCategory.Validation,
        defaultNextSteps: ["Check the expected format for this flag", "Run with --help for usage information"],
    },
    FILE_NOT_FOUND: {
        pattern: "File not found: {{path}}",
        category: ErrorCategory.Configuration,
        defaultNextSteps: ["Verify the file path is correct", "Check that the file exists and is readable"],
    },
    PROCESS_NOT_FOUND: {
        pattern: "Process entry file not found: {{path}}",
        category: ErrorCategory.Configuration,
        defaultNextSteps: [
            "Ensure the path is correct and points to a valid JS/TS module",
            "Check that the file has been compiled if using TypeScript",
        ],
    },
    EXPORT_NOT_FOUND: {
        pattern: "Process module {{path}} does not export '{{exportName}}'",
        category: ErrorCategory.Configuration,
        defaultNextSteps: [
            "Check available exports in your module",
            "Use --entry {{path}}#default for default export",
        ],
        defaultSuggestions: ["Did you mean to use a different export name?"],
    },
    EFFECT_NOT_FOUND: {
        pattern: "Effect {{effectId}} not found at {{runDir}}",
        category: ErrorCategory.Validation,
        defaultNextSteps: ["Verify the effect ID is correct", "Run task:list to see available effects"],
    },
    EFFECT_WRONG_STATUS: {
        pattern: "Effect {{effectId}} is not {{expectedStatus}} (current status={{actualStatus}})",
        category: ErrorCategory.Validation,
        defaultNextSteps: ["Check the current effect status with task:show", "Ensure you're operating on the correct effect"],
    },
    RUN_NOT_FOUND: {
        pattern: "Run directory not found: {{runDir}}",
        category: ErrorCategory.Configuration,
        defaultNextSteps: ["Verify the run directory path", "Ensure the run has been created with run:create"],
    },
    JSON_PARSE_ERROR: {
        pattern: "Failed to parse {{file}} as JSON: {{error}}",
        category: ErrorCategory.Validation,
        defaultNextSteps: ["Check that the file contains valid JSON", "Validate JSON syntax with a linter"],
    },
    MISSING_PROCESS_CONTEXT: {
        pattern: "No active process context found on the current async call stack",
        category: ErrorCategory.Runtime,
        defaultNextSteps: [
            "Ensure you are calling this from within a babysitter process function",
            "Check that async context is properly maintained",
        ],
    },
    INVOCATION_COLLISION: {
        pattern: "Invocation key {{invocationKey}} is already in use within this run",
        category: ErrorCategory.Runtime,
        defaultNextSteps: [
            "Ensure unique invocation keys for each task invocation",
            "Check for duplicate task calls in your process",
        ],
    },
};
/**
 * Interpolates context variables into a template pattern
 */
function interpolateTemplate(pattern, context) {
    return pattern.replace(/\{\{(\w+)\}\}/g, (match, key) => {
        const value = context[key];
        if (value === undefined || value === null) {
            return match; // Keep placeholder if no value provided
        }
        return String(value);
    });
}
/**
 * Creates an error message from a template and context
 */
function createErrorMessage(templateKey, context) {
    const template = exports.ERROR_TEMPLATES[templateKey];
    return interpolateTemplate(template.pattern, context);
}
// ============================================================================
// "Did You Mean?" Suggestions
// ============================================================================
/**
 * Common typos and their corrections for CLI commands
 */
const COMMAND_TYPOS = {
    "run:create": ["run:creat", "run:craete", "runcreate", "create:run", "run-create"],
    "run:status": ["run:stat", "run:staus", "runstatus", "status:run", "run-status"],
    "run:iterate": ["run:iter", "run:itterate", "runiterate", "iterate:run", "run-iterate"],
    "run:events": ["run:event", "runevents", "events:run", "run-events"],
    "run:rebuild-state": ["run:rebuild", "run:rebuildstate", "rebuild-state", "run-rebuild-state"],
    "run:repair-journal": ["run:repair", "run:repairjournal", "repair-journal", "run-repair-journal"],
    "task:post": ["task:pst", "taskpost", "post:task", "task-post"],
    "task:list": ["task:lst", "tasklist", "list:task", "task-list", "tasks:list"],
    "task:show": ["task:shw", "taskshow", "show:task", "task-show"],
};
/**
 * Common typos and their corrections for CLI flags
 */
const FLAG_TYPOS = {
    "--runs-dir": ["--runsdir", "--run-dir", "--rundir", "-runs-dir"],
    "--process-id": ["--processid", "--process_id", "-process-id"],
    "--entry": ["--enrty", "--entyr", "-entry"],
    "--inputs": ["--input", "--inpust", "-inputs"],
    "--json": ["--JSON", "-json", "--jsn"],
    "--dry-run": ["--dryrun", "--dry_run", "-dry-run"],
    "--verbose": ["--verbos", "-verbose", "--vebrose"],
    "--status": ["--staus", "--stats", "-status"],
    "--pending": ["--peding", "-pending"],
};
/**
 * Computes Levenshtein distance between two strings
 */
function levenshteinDistance(a, b) {
    const matrix = [];
    for (let i = 0; i <= b.length; i++) {
        matrix[i] = [i];
    }
    for (let j = 0; j <= a.length; j++) {
        matrix[0][j] = j;
    }
    for (let i = 1; i <= b.length; i++) {
        for (let j = 1; j <= a.length; j++) {
            if (b.charAt(i - 1) === a.charAt(j - 1)) {
                matrix[i][j] = matrix[i - 1][j - 1];
            }
            else {
                matrix[i][j] = Math.min(matrix[i - 1][j - 1] + 1, matrix[i][j - 1] + 1, matrix[i - 1][j] + 1);
            }
        }
    }
    return matrix[b.length][a.length];
}
/**
 * Suggests corrections for a potentially misspelled command
 */
function suggestCommand(input) {
    const normalizedInput = input.toLowerCase().trim();
    // Check direct typo mappings first
    for (const [correct, typos] of Object.entries(COMMAND_TYPOS)) {
        if (typos.includes(normalizedInput)) {
            return correct;
        }
    }
    // Fall back to Levenshtein distance
    const commands = Object.keys(COMMAND_TYPOS);
    let bestMatch;
    let bestDistance = Infinity;
    for (const command of commands) {
        const distance = levenshteinDistance(normalizedInput, command);
        if (distance < bestDistance && distance <= 3) {
            bestDistance = distance;
            bestMatch = command;
        }
    }
    return bestMatch;
}
/**
 * Suggests corrections for a potentially misspelled flag
 */
function suggestFlag(input) {
    const normalizedInput = input.toLowerCase().trim();
    // Check direct typo mappings first
    for (const [correct, typos] of Object.entries(FLAG_TYPOS)) {
        if (typos.includes(normalizedInput)) {
            return correct;
        }
    }
    // Fall back to Levenshtein distance
    const flags = Object.keys(FLAG_TYPOS);
    let bestMatch;
    let bestDistance = Infinity;
    for (const flag of flags) {
        const distance = levenshteinDistance(normalizedInput, flag);
        if (distance < bestDistance && distance <= 3) {
            bestDistance = distance;
            bestMatch = flag;
        }
    }
    return bestMatch;
}
/**
 * Generic suggestion helper that finds the closest match from a list
 */
function suggestFix(input, validOptions, maxDistance = 3) {
    const normalizedInput = input.toLowerCase().trim();
    let bestMatch;
    let bestDistance = Infinity;
    for (const option of validOptions) {
        const distance = levenshteinDistance(normalizedInput, option.toLowerCase());
        if (distance < bestDistance && distance <= maxDistance) {
            bestDistance = distance;
            bestMatch = option;
        }
    }
    return bestMatch;
}
// ============================================================================
// Core Error Classes
// ============================================================================
class BabysitterRuntimeError extends Error {
    details;
    category;
    suggestions;
    nextSteps;
    constructor(name, message, options) {
        super(message);
        this.name = name;
        // Handle backward compatibility: if options looks like old-style details object
        if (options && !isErrorOptions(options)) {
            this.details = options;
            this.category = ErrorCategory.Runtime;
            this.suggestions = [];
            this.nextSteps = [];
        }
        else {
            const opts = options;
            this.details = opts?.details;
            this.category = opts?.category ?? ErrorCategory.Runtime;
            this.suggestions = opts?.suggestions ?? [];
            this.nextSteps = opts?.nextSteps ?? [];
            if (opts?.cause) {
                this.cause = opts.cause;
            }
        }
    }
    /**
     * Creates an error from a template
     */
    static fromTemplate(name, templateKey, context, additionalOptions) {
        const template = exports.ERROR_TEMPLATES[templateKey];
        const message = interpolateTemplate(template.pattern, context);
        return new BabysitterRuntimeError(name, message, {
            category: additionalOptions?.category ?? template.category,
            suggestions: additionalOptions?.suggestions ?? template.defaultSuggestions ?? [],
            nextSteps: additionalOptions?.nextSteps ?? template.defaultNextSteps ?? [],
            details: { ...context, ...(additionalOptions?.details ?? {}) },
            cause: additionalOptions?.cause,
        });
    }
}
exports.BabysitterRuntimeError = BabysitterRuntimeError;
/**
 * Type guard to check if an object is BabysitterErrorOptions vs legacy details
 */
function isErrorOptions(obj) {
    if (!obj || typeof obj !== "object")
        return false;
    const keys = Object.keys(obj);
    const optionKeys = ["category", "suggestions", "nextSteps", "details", "cause"];
    return keys.some((key) => optionKeys.includes(key));
}
class BabysitterIntrinsicError extends BabysitterRuntimeError {
    isIntrinsic = true;
    constructor(name, message, options) {
        super(name, message, options);
    }
}
exports.BabysitterIntrinsicError = BabysitterIntrinsicError;
class EffectRequestedError extends BabysitterIntrinsicError {
    action;
    constructor(action) {
        super("EffectRequestedError", `Effect ${action.effectId} requested`, { details: { action } });
        this.action = action;
    }
}
exports.EffectRequestedError = EffectRequestedError;
class EffectPendingError extends BabysitterIntrinsicError {
    action;
    constructor(action) {
        super("EffectPendingError", `Effect ${action.effectId} pending`, { details: { action } });
        this.action = action;
    }
}
exports.EffectPendingError = EffectPendingError;
class ParallelPendingError extends BabysitterIntrinsicError {
    batch;
    effects;
    constructor(batch) {
        super("ParallelPendingError", "One or more parallel invocations are pending", {
            details: {
                payload: (0, batching_1.toParallelPendingPayload)(batch),
                effects: batch.actions,
            },
        });
        this.batch = batch;
        this.effects = batch.actions;
    }
}
exports.ParallelPendingError = ParallelPendingError;
class InvocationCollisionError extends BabysitterRuntimeError {
    invocationKey;
    constructor(invocationKey) {
        super("InvocationCollisionError", `Invocation key ${invocationKey} is already in use within this run`, {
            category: ErrorCategory.Runtime,
            details: { invocationKey },
            nextSteps: [
                "Ensure unique invocation keys for each task invocation",
                "Check for duplicate task calls in your process",
            ],
        });
        this.invocationKey = invocationKey;
    }
}
exports.InvocationCollisionError = InvocationCollisionError;
class RunFailedError extends BabysitterRuntimeError {
    constructor(message, options) {
        super("RunFailedError", message, options);
    }
}
exports.RunFailedError = RunFailedError;
class MissingProcessContextError extends BabysitterRuntimeError {
    constructor() {
        super("MissingProcessContextError", "No active process context found on the current async call stack", {
            category: ErrorCategory.Runtime,
            nextSteps: [
                "Ensure you are calling this from within a babysitter process function",
                "Check that async context is properly maintained",
            ],
        });
    }
}
exports.MissingProcessContextError = MissingProcessContextError;
class InvalidTaskDefinitionError extends BabysitterRuntimeError {
    constructor(reason) {
        super("InvalidTaskDefinitionError", reason, {
            category: ErrorCategory.Validation,
            nextSteps: ["Review the task definition requirements", "Check the task schema documentation"],
        });
    }
}
exports.InvalidTaskDefinitionError = InvalidTaskDefinitionError;
class InvalidSleepTargetError extends BabysitterRuntimeError {
    constructor(value) {
        super("InvalidSleepTargetError", `Invalid sleep target: ${value}`, {
            category: ErrorCategory.Validation,
            nextSteps: [
                "Provide a valid duration (positive number of milliseconds)",
                "Or provide a valid Date object or ISO 8601 date string",
            ],
        });
    }
}
exports.InvalidSleepTargetError = InvalidSleepTargetError;
/**
 * Formats an error with full context including category, suggestions, and next steps
 */
function formatErrorWithContext(error, options = {}) {
    const { colors = false, includeStack = false, prefix = "" } = options;
    const lines = [];
    // Color helpers (no-op if colors disabled)
    const red = colors ? (s) => `\x1b[31m${s}\x1b[0m` : (s) => s;
    const yellow = colors ? (s) => `\x1b[33m${s}\x1b[0m` : (s) => s;
    const cyan = colors ? (s) => `\x1b[36m${s}\x1b[0m` : (s) => s;
    const dim = colors ? (s) => `\x1b[2m${s}\x1b[0m` : (s) => s;
    const bold = colors ? (s) => `\x1b[1m${s}\x1b[0m` : (s) => s;
    // Main error line
    const isBabysitterError = error instanceof BabysitterRuntimeError;
    const category = isBabysitterError ? error.category : ErrorCategory.Internal;
    const categoryLabel = exports.ERROR_CATEGORY_DESCRIPTIONS[category];
    lines.push(`${prefix}${red(bold("Error:"))} ${error.message}`);
    lines.push(`${prefix}${dim(`[${error.name}] Category: ${categoryLabel}`)}`);
    // Suggestions ("Did you mean?")
    if (isBabysitterError && error.suggestions.length > 0) {
        lines.push("");
        lines.push(`${prefix}${yellow("Did you mean?")}`);
        for (const suggestion of error.suggestions) {
            lines.push(`${prefix}  - ${suggestion}`);
        }
    }
    // Next steps
    if (isBabysitterError && error.nextSteps.length > 0) {
        lines.push("");
        lines.push(`${prefix}${cyan("Next Steps:")}`);
        for (const step of error.nextSteps) {
            lines.push(`${prefix}  - ${step}`);
        }
    }
    // Stack trace
    if (includeStack && error.stack) {
        lines.push("");
        lines.push(`${prefix}${dim("Stack trace:")}`);
        const stackLines = error.stack.split("\n").slice(1); // Skip the error message line
        for (const stackLine of stackLines) {
            lines.push(`${prefix}${dim(stackLine)}`);
        }
    }
    return lines.join("\n");
}
/**
 * Formats next steps as a bulleted list
 */
function formatNextSteps(nextSteps, options = {}) {
    const { prefix = "", colors = false } = options;
    const cyan = colors ? (s) => `\x1b[36m${s}\x1b[0m` : (s) => s;
    if (nextSteps.length === 0) {
        return "";
    }
    const lines = [`${prefix}${cyan("Next Steps:")}`];
    for (const step of nextSteps) {
        lines.push(`${prefix}  - ${step}`);
    }
    return lines.join("\n");
}
/**
 * Converts an error to a structured JSON-serializable format
 */
function toStructuredError(error, includeStack = false) {
    const isBabysitterError = error instanceof BabysitterRuntimeError;
    return {
        name: error.name,
        message: error.message,
        category: isBabysitterError ? error.category : ErrorCategory.Internal,
        categoryDescription: exports.ERROR_CATEGORY_DESCRIPTIONS[isBabysitterError ? error.category : ErrorCategory.Internal],
        suggestions: isBabysitterError ? error.suggestions : [],
        nextSteps: isBabysitterError ? error.nextSteps : [],
        details: isBabysitterError ? error.details : undefined,
        stack: includeStack ? error.stack : undefined,
    };
}
// ============================================================================
// Utility Functions
// ============================================================================
function isIntrinsicError(error) {
    return Boolean(error && typeof error === "object" && error.isIntrinsic);
}
function isBabysitterError(error) {
    return error instanceof BabysitterRuntimeError;
}
function rehydrateSerializedError(data) {
    const name = data?.name ?? "TaskError";
    const message = data?.message ?? "Task failed";
    const err = new Error(message);
    err.name = name;
    if (data?.stack) {
        err.stack = data.stack;
    }
    if (data?.data !== undefined) {
        err.data = data.data;
    }
    return err;
}
