// @a5c-ai/babysitter-sdk/dist/runtime/intrinsics/breakpoint.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.runBreakpointIntrinsic = runBreakpointIntrinsic;
const task_1 = require("./task");
const BREAKPOINT_TASK_ID = "__sdk.breakpoint";
const DEFAULT_BREAKPOINT_LABEL = "breakpoint";
const breakpointTask = {
    id: BREAKPOINT_TASK_ID,
    build(args) {
        return {
            kind: "breakpoint",
            title: args.label,
            metadata: {
                payload: args?.payload,
                requestedAt: args.requestedAt,
                label: args.label,
            },
        };
    },
};
function runBreakpointIntrinsic(payload, context, options) {
    const label = deriveBreakpointLabel(payload, options?.label);
    const invokeOptions = { ...options, label };
    return (0, task_1.runTaskIntrinsic)({
        task: breakpointTask,
        args: { payload, label, requestedAt: context.now().toISOString() },
        invokeOptions,
        context,
    });
}
function deriveBreakpointLabel(payload, provided) {
    if (typeof provided === "string" && provided.length) {
        return provided;
    }
    if (payload && typeof payload === "object" && "label" in payload) {
        const inferred = payload.label;
        if (typeof inferred === "string" && inferred.length) {
            return inferred;
        }
    }
    return DEFAULT_BREAKPOINT_LABEL;
}
