// @a5c-ai/babysitter-sdk/dist/profiles/types.js
"use strict";
/**
 * Profile Types for Babysitter SDK
 *
 * This module defines the TypeScript interfaces for user and project profiles.
 * Profiles store persistent configuration, preferences, and contextual information
 * that the babysitter runtime uses to tailor its behavior.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PROJECT_PROFILE_MD_FILENAME = exports.PROJECT_PROFILE_FILENAME = exports.USER_PROFILE_MD_FILENAME = exports.USER_PROFILE_FILENAME = exports.PROJECT_PROFILE_SCHEMA_VERSION = exports.USER_PROFILE_SCHEMA_VERSION = void 0;
/**
 * Schema version for user profile serialization.
 * Used to detect and migrate older profile formats.
 */
exports.USER_PROFILE_SCHEMA_VERSION = "2026.01.user-profile-v1";
/**
 * Schema version for project profile serialization.
 * Used to detect and migrate older project profile formats.
 */
exports.PROJECT_PROFILE_SCHEMA_VERSION = "2026.01.project-profile-v1";
/**
 * Filename for the JSON-serialized user profile.
 */
exports.USER_PROFILE_FILENAME = "user-profile.json";
/**
 * Filename for the human-readable markdown user profile summary.
 */
exports.USER_PROFILE_MD_FILENAME = "user-profile.md";
/**
 * Filename for the JSON-serialized project profile.
 */
exports.PROJECT_PROFILE_FILENAME = "project-profile.json";
/**
 * Filename for the human-readable markdown project profile summary.
 */
exports.PROJECT_PROFILE_MD_FILENAME = "project-profile.md";
