// @a5c-ai/babysitter-sdk/dist/cli/main.js
#!/usr/bin/env node
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports._collapseDoubledA5cRuns = void 0;
exports._resolveRunDir = resolveRunDir;
exports.createBabysitterCli = createBabysitterCli;
const node_fs_1 = require("node:fs");
const path = __importStar(require("node:path"));
const node_url_1 = require("node:url");
const crypto = __importStar(require("node:crypto"));
const resolveInputPath_1 = require("./resolveInputPath");
const commitEffectResult_1 = require("../runtime/commitEffectResult");
const createRun_1 = require("../runtime/createRun");
const effectIndex_1 = require("../runtime/replay/effectIndex");
const stateCache_1 = require("../runtime/replay/stateCache");
const ulids_1 = require("../storage/ulids");
const tasks_1 = require("../storage/tasks");
const journal_1 = require("../storage/journal");
const runFiles_1 = require("../storage/runFiles");
const runIterate_1 = require("./commands/runIterate");
const runExecuteTasks_1 = require("./commands/runExecuteTasks");
const health_1 = require("./commands/health");
const configure_1 = require("./commands/configure");
const session_1 = require("./commands/session");
const skill_1 = require("./commands/skill");
const hookLog_1 = require("./commands/hookLog");
const hookRun_1 = require("./commands/hookRun");
const profile_1 = require("./commands/profile");
const completionProof_1 = require("./completionProof");
const harness_1 = require("../harness");
const exceptions_1 = require("../runtime/exceptions");
const defaults_1 = require("../config/defaults");
const USAGE = `Usage:
  babysitter run:create --process-id <id> --entry <path#export> [--runs-dir <dir>] [--inputs <file>] [--run-id <id>] [--process-revision <rev>] [--request <id>] [--prompt <text>] [--harness <name>] [--session-id <id>] [--plugin-root <dir>] [--json] [--dry-run]
  babysitter run:status <runDir> [--runs-dir <dir>] [--json]
  babysitter run:events <runDir> [--runs-dir <dir>] [--json] [--limit <n>] [--reverse] [--filter-type <type>]
  babysitter run:rebuild-state <runDir> [--runs-dir <dir>] [--json] [--dry-run]
  babysitter run:repair-journal <runDir> [--runs-dir <dir>] [--json] [--dry-run]
  babysitter run:iterate <runDir> [--runs-dir <dir>] [--json] [--verbose] [--iteration <n>]
  babysitter run:execute-tasks <runDir> [--runs-dir <dir>] [--json] [--verbose] [--dry-run] [--max-tasks <n>] [--kind <kind>] [--timeout <ms>]
  babysitter task:post <runDir> <effectId> --status <ok|error> [--runs-dir <dir>] [--json] [--dry-run] [--value <file>] [--error <file>] [--stdout-ref <ref>] [--stderr-ref <ref>] [--stdout-file <file>] [--stderr-file <file>] [--started-at <iso8601>] [--finished-at <iso8601>] [--metadata <file>] [--invocation-key <key>]
  babysitter task:list <runDir> [--runs-dir <dir>] [--pending] [--kind <kind>] [--json]
  babysitter task:show <runDir> <effectId> [--runs-dir <dir>] [--json]
  babysitter session:init --session-id <id> --state-dir <dir> [--max-iterations <n>] [--run-id <id>] [--prompt <text>] [--json]
  babysitter session:associate --session-id <id> --state-dir <dir> --run-id <id> [--json]
  babysitter session:resume --session-id <id> --state-dir <dir> --run-id <id> [--max-iterations <n>] [--runs-dir <dir>] [--json]
  babysitter session:state --session-id <id> --state-dir <dir> [--json]
  babysitter session:update --session-id <id> --state-dir <dir> [--iteration <n>] [--last-iteration-at <iso8601>] [--iteration-times <csv>] [--delete] [--json]
  babysitter session:check-iteration --session-id <id> --state-dir <dir> [--json]
  babysitter session:last-message --transcript-path <file> [--json]
  babysitter session:iteration-message --iteration <n> [--run-id <id>] [--runs-dir <dir>] [--plugin-root <dir>] [--json]
  babysitter skill:discover --plugin-root <dir> [--run-id <id>] [--cache-ttl <seconds>] [--runs-dir <dir>] [--include-remote] [--summary-only] [--process-path <path>] [--json]
  babysitter hook:log --hook-type <type> --log-file <path> [--json]
  babysitter hook:run --hook-type <stop|session-start> [--harness <claude-code>] [--plugin-root <dir>] [--state-dir <dir>] [--runs-dir <dir>] [--json] [--verbose]
  babysitter skill:fetch-remote --source-type <github|well-known> --url <url> [--json]
  babysitter profile:read --user|--project [--dir <dir>] [--json]
  babysitter profile:write --user|--project --input <file> [--dir <dir>] [--json]
  babysitter profile:merge --user|--project --input <file> [--dir <dir>] [--json]
  babysitter profile:render --user|--project [--dir <dir>] [--json]
  babysitter health [--json] [--verbose]
  babysitter configure [show|validate|paths] [--json] [--defaults-only]
  babysitter version

Global flags:
  --runs-dir <dir>   Override the runs directory (defaults to .a5c/runs).
  --json             Emit JSON output when supported by the command.
  --dry-run          Describe planned mutations without changing on-disk state.
  --verbose          Log resolved paths and options to stderr for debugging.
  --show-config      Show current configuration before executing command.
  --help, -h         Show this help text.
  --version, -v      Show CLI version.`;
/**
 * Maximum size for inline result preview.
 * @see DEFAULTS.largeResultPreviewLimit for the centralized default
 */
const LARGE_RESULT_PREVIEW_LIMIT = defaults_1.DEFAULTS.largeResultPreviewLimit;
function parseArgs(argv) {
    const [initialCommand, ...rest] = argv;
    const parsed = {
        command: initialCommand,
        runsDir: process.env[defaults_1.CONFIG_ENV_VARS.RUNS_DIR] ?? defaults_1.DEFAULTS.runsDir,
        json: false,
        dryRun: false,
        verbose: false,
        helpRequested: false,
        pendingOnly: false,
        reverseOrder: false,
        showConfig: false,
        defaultsOnly: false,
    };
    if (parsed.command === "--help" || parsed.command === "-h") {
        parsed.command = undefined;
        parsed.helpRequested = true;
    }
    if (parsed.command === "--version" || parsed.command === "-v") {
        parsed.command = "version";
    }
    const positionals = [];
    for (let i = 0; i < rest.length; i += 1) {
        const arg = rest[i];
        if (arg === "--help" || arg === "-h") {
            parsed.helpRequested = true;
            continue;
        }
        if (arg === "--version" || arg === "-v") {
            parsed.command = "version";
            continue;
        }
        if (arg === "--runs-dir") {
            parsed.runsDir = expectFlagValue(rest, ++i, "--runs-dir");
            continue;
        }
        if (arg === "--json") {
            parsed.json = true;
            continue;
        }
        if (arg === "--dry-run") {
            parsed.dryRun = true;
            continue;
        }
        if (arg === "--verbose") {
            parsed.verbose = true;
            continue;
        }
        if (arg === "--pending") {
            parsed.pendingOnly = true;
            continue;
        }
        if (arg === "--kind") {
            parsed.kindFilter = expectFlagValue(rest, ++i, "--kind");
            continue;
        }
        if (arg === "--limit") {
            const raw = expectFlagValue(rest, ++i, "--limit");
            parsed.limit = parsePositiveInteger(raw, "--limit");
            continue;
        }
        if (arg === "--iteration") {
            const raw = expectFlagValue(rest, ++i, "--iteration");
            parsed.iteration = parsePositiveInteger(raw, "--iteration");
            continue;
        }
        if (arg === "--reverse") {
            parsed.reverseOrder = true;
            continue;
        }
        if (arg === "--filter-type") {
            parsed.filterType = expectFlagValue(rest, ++i, "--filter-type");
            continue;
        }
        if (arg === "--status") {
            const raw = expectFlagValue(rest, ++i, "--status");
            const normalized = raw.toLowerCase();
            if (normalized !== "ok" && normalized !== "error") {
                throw new Error(`--status must be "ok" or "error" (received: ${raw})`);
            }
            parsed.taskStatus = normalized === "ok" ? "ok" : "error";
            continue;
        }
        if (arg === "--value") {
            parsed.valuePath = expectFlagValue(rest, ++i, "--value");
            continue;
        }
        if (arg === "--error") {
            parsed.errorPath = expectFlagValue(rest, ++i, "--error");
            continue;
        }
        if (arg === "--stdout-ref") {
            parsed.stdoutRef = expectFlagValue(rest, ++i, "--stdout-ref");
            continue;
        }
        if (arg === "--stderr-ref") {
            parsed.stderrRef = expectFlagValue(rest, ++i, "--stderr-ref");
            continue;
        }
        if (arg === "--stdout-file") {
            parsed.stdoutFile = expectFlagValue(rest, ++i, "--stdout-file");
            continue;
        }
        if (arg === "--stderr-file") {
            parsed.stderrFile = expectFlagValue(rest, ++i, "--stderr-file");
            continue;
        }
        if (arg === "--started-at") {
            parsed.startedAt = expectFlagValue(rest, ++i, "--started-at");
            continue;
        }
        if (arg === "--finished-at") {
            parsed.finishedAt = expectFlagValue(rest, ++i, "--finished-at");
            continue;
        }
        if (arg === "--metadata") {
            parsed.metadataPath = expectFlagValue(rest, ++i, "--metadata");
            continue;
        }
        if (arg === "--invocation-key") {
            parsed.invocationKey = expectFlagValue(rest, ++i, "--invocation-key");
            continue;
        }
        if (arg === "--process-id") {
            parsed.processId = expectFlagValue(rest, ++i, "--process-id");
            continue;
        }
        if (arg === "--entry") {
            parsed.entrySpecifier = expectFlagValue(rest, ++i, "--entry");
            continue;
        }
        if (arg === "--inputs") {
            parsed.inputsPath = expectFlagValue(rest, ++i, "--inputs");
            continue;
        }
        if (arg === "--run-id") {
            parsed.runIdOverride = expectFlagValue(rest, ++i, "--run-id");
            continue;
        }
        if (arg === "--process-revision") {
            parsed.processRevision = expectFlagValue(rest, ++i, "--process-revision");
            continue;
        }
        if (arg === "--request") {
            parsed.requestId = expectFlagValue(rest, ++i, "--request");
            continue;
        }
        if (arg === "--show-config") {
            parsed.showConfig = true;
            continue;
        }
        if (arg === "--defaults-only") {
            parsed.defaultsOnly = true;
            continue;
        }
        // Session command flags
        if (arg === "--session-id") {
            parsed.sessionId = expectFlagValue(rest, ++i, "--session-id");
            continue;
        }
        if (arg === "--state-dir") {
            parsed.stateDir = expectFlagValue(rest, ++i, "--state-dir");
            continue;
        }
        if (arg === "--max-iterations") {
            const raw = expectFlagValue(rest, ++i, "--max-iterations");
            parsed.maxIterations = parsePositiveInteger(raw, "--max-iterations");
            continue;
        }
        if (arg === "--prompt") {
            parsed.prompt = expectFlagValue(rest, ++i, "--prompt");
            continue;
        }
        if (arg === "--last-iteration-at") {
            parsed.lastIterationAt = expectFlagValue(rest, ++i, "--last-iteration-at");
            continue;
        }
        if (arg === "--iteration-times") {
            parsed.iterationTimes = expectFlagValue(rest, ++i, "--iteration-times");
            continue;
        }
        if (arg === "--max-tasks") {
            const raw = expectFlagValue(rest, ++i, "--max-tasks");
            parsed.maxTasks = parsePositiveInteger(raw, "--max-tasks");
            continue;
        }
        if (arg === "--timeout") {
            const raw = expectFlagValue(rest, ++i, "--timeout");
            parsed.timeout = parsePositiveInteger(raw, "--timeout");
            continue;
        }
        if (arg === "--delete") {
            parsed.deleteSession = true;
            continue;
        }
        if (arg === "--transcript-path") {
            parsed.transcriptPath = expectFlagValue(rest, ++i, "--transcript-path");
            continue;
        }
        // Hook command flags
        if (arg === "--hook-type") {
            parsed.hookType = expectFlagValue(rest, ++i, "--hook-type");
            continue;
        }
        if (arg === "--harness") {
            parsed.harness = expectFlagValue(rest, ++i, "--harness");
            continue;
        }
        if (arg === "--log-file") {
            parsed.logFile = expectFlagValue(rest, ++i, "--log-file");
            continue;
        }
        // Skill command flags
        if (arg === "--plugin-root") {
            parsed.pluginRoot = expectFlagValue(rest, ++i, "--plugin-root");
            continue;
        }
        if (arg === "--cache-ttl") {
            const raw = expectFlagValue(rest, ++i, "--cache-ttl");
            parsed.cacheTtl = parsePositiveInteger(raw, "--cache-ttl");
            continue;
        }
        if (arg === "--source-type") {
            const raw = expectFlagValue(rest, ++i, "--source-type");
            if (raw !== "github" && raw !== "well-known") {
                throw new Error(`--source-type must be "github" or "well-known" (received: ${raw})`);
            }
            parsed.sourceType = raw;
            continue;
        }
        if (arg === "--url") {
            parsed.url = expectFlagValue(rest, ++i, "--url");
            continue;
        }
        if (arg === "--include-remote") {
            parsed.includeRemote = true;
            continue;
        }
        if (arg === "--summary-only") {
            parsed.summaryOnly = true;
            continue;
        }
        if (arg === "--process-path") {
            parsed.processPath = expectFlagValue(rest, ++i, "--process-path");
            continue;
        }
        // Profile command flags
        if (arg === "--user") {
            parsed.profileUser = true;
            continue;
        }
        if (arg === "--project") {
            parsed.profileProject = true;
            continue;
        }
        if (arg === "--input") {
            parsed.profileInputPath = expectFlagValue(rest, ++i, "--input");
            continue;
        }
        if (arg === "--dir") {
            parsed.profileDir = expectFlagValue(rest, ++i, "--dir");
            continue;
        }
        positionals.push(arg);
    }
    if (parsed.command === "task:post") {
        [parsed.runDirArg, parsed.effectId] = positionals;
    }
    else if (parsed.command === "task:list") {
        [parsed.runDirArg] = positionals;
    }
    else if (parsed.command === "task:show") {
        [parsed.runDirArg, parsed.effectId] = positionals;
    }
    else if (parsed.command === "run:status") {
        [parsed.runDirArg] = positionals;
    }
    else if (parsed.command === "run:iterate") {
        [parsed.runDirArg] = positionals;
    }
    else if (parsed.command === "run:events") {
        [parsed.runDirArg] = positionals;
    }
    else if (parsed.command === "run:execute-tasks") {
        [parsed.runDirArg] = positionals;
    }
    else if (parsed.command === "run:rebuild-state") {
        [parsed.runDirArg] = positionals;
    }
    else if (parsed.command === "run:repair-journal") {
        [parsed.runDirArg] = positionals;
    }
    else if (parsed.command === "configure") {
        [parsed.configureSubcommand] = positionals;
    }
    return parsed;
}
/**
 * Resolve a run directory from a base directory and a user-provided argument.
 *
 * Handles several common edge cases:
 *  - Absolute paths are used directly (no base dir prepended).
 *  - If the arg already contains the base dir prefix (e.g. ".a5c/runs/01RUN"
 *    when base is ".a5c/runs"), it is resolved from CWD to avoid doubling.
 *  - Doubled ".a5c/runs" segments in the final path are collapsed.
 *  - If the resolved path doesn't exist, falls back to resolving from CWD.
 */
function resolveRunDir(baseDir, runDirArg) {
    if (!runDirArg)
        throw new Error("Run directory argument is required.");
    // Absolute path → use directly
    if (path.isAbsolute(runDirArg)) {
        return collapseDoubledA5cRuns(path.normalize(runDirArg));
    }
    // Detect if arg already starts with the base dir prefix to avoid
    // ".a5c/runs" + ".a5c/runs/01RUN" → ".a5c/runs/.a5c/runs/01RUN"
    const normalBase = normalizePosix(baseDir);
    const normalArg = normalizePosix(runDirArg);
    if (normalBase && (normalArg === normalBase || normalArg.startsWith(normalBase + "/"))) {
        return collapseDoubledA5cRuns(path.resolve(runDirArg));
    }
    // Standard resolution: baseDir + arg
    const standard = collapseDoubledA5cRuns(path.resolve(baseDir, runDirArg));
    // Fallback: if the standard path doesn't exist, try from CWD
    try {
        if (!(0, node_fs_1.existsSync)(standard)) {
            const fromCwd = path.resolve(runDirArg);
            if ((0, node_fs_1.existsSync)(fromCwd))
                return collapseDoubledA5cRuns(fromCwd);
        }
    }
    catch {
        // Ignore filesystem errors during resolution
    }
    return standard;
}
/** Normalize a path to forward slashes with no trailing slash (for prefix comparison). */
function normalizePosix(p) {
    return path.normalize(p).replace(/\\/g, "/").replace(/\/+$/, "");
}
/** Collapse doubled ".a5c/runs" segments — delegates to shared utility. */
const collapseDoubledA5cRuns = resolveInputPath_1.collapseDoubledA5cRuns;
exports._collapseDoubledA5cRuns = collapseDoubledA5cRuns;
function expectFlagValue(args, index, flag) {
    const value = args[index];
    if (!value) {
        throw new Error(`${flag} requires a value`);
    }
    return value;
}
function parsePositiveInteger(raw, flag) {
    const parsed = Number(raw);
    if (!Number.isFinite(parsed) || parsed <= 0) {
        throw new Error(`${flag} must be a positive integer`);
    }
    return Math.floor(parsed);
}
function summarizeActions(actions) {
    return actions.map((action) => ({
        effectId: action.effectId,
        kind: action.kind,
        label: action.label,
    }));
}
function _logPendingActions(actions, options = {}) {
    const summaries = summarizeActions(actions);
    if (options.command && options.includeHeader !== false) {
        const headerParts = [
            `[${options.command}] status=waiting`,
            `pending=${summaries.length}`,
            ...(options.metadataParts ?? []),
        ];
        console.error(headerParts.join(" "));
    }
    for (const summary of summaries) {
        const label = summary.label ? ` ${summary.label}` : "";
        console.error(`- ${summary.effectId} [${summary.kind}]${label}`);
    }
    return summaries;
}
function countActionsByKind(actions) {
    const counts = new Map();
    for (const action of actions) {
        counts.set(action.kind, (counts.get(action.kind) ?? 0) + 1);
    }
    return Object.fromEntries(Array.from(counts.entries()).sort(([a], [b]) => a.localeCompare(b)));
}
function _enrichIterationMetadata(metadata, pendingActions) {
    if (!pendingActions?.length) {
        return metadata;
    }
    if (metadata?.pendingEffectsByKind) {
        return metadata;
    }
    return {
        ...(metadata ?? {}),
        pendingEffectsByKind: countActionsByKind(pendingActions),
    };
}
function _logSleepHints(command, actions) {
    for (const action of actions) {
        const sleepMs = action.schedulerHints?.sleepUntilEpochMs;
        if (typeof sleepMs !== "number")
            continue;
        const iso = new Date(sleepMs).toISOString();
        const label = action.label ? ` ${action.label}` : "";
        const pendingInfo = typeof action.schedulerHints?.pendingCount === "number"
            ? ` pendingCount=${action.schedulerHints.pendingCount}`
            : "";
        console.error(`[${command}] sleep-until=${iso} effect=${action.effectId}${label}${pendingInfo}`);
    }
}
function formatResolvedEntrypoint(importPath, exportName) {
    return `${importPath}${exportName ? `#${exportName}` : ""}`;
}
function formatVerboseValue(value) {
    if (value === null)
        return "null";
    if (value === undefined)
        return "undefined";
    if (typeof value === "string")
        return value;
    if (typeof value === "number" || typeof value === "boolean")
        return String(value);
    return JSON.stringify(value);
}
function allowSecretLogs(parsed) {
    if (!parsed.json || !parsed.verbose) {
        return false;
    }
    const raw = process.env.BABYSITTER_ALLOW_SECRET_LOGS;
    if (!raw) {
        return false;
    }
    const normalized = raw.trim().toLowerCase();
    return normalized === "1" || normalized === "true";
}
function logVerbose(command, parsed, details) {
    if (!parsed.verbose)
        return;
    const formatted = Object.entries(details)
        .filter(([, value]) => value !== undefined)
        .map(([key, value]) => `${key}=${formatVerboseValue(value)}`)
        .join(" ");
    console.error(`[${command}] verbose ${formatted}`);
}
function isJsonRecord(value) {
    return Boolean(value) && typeof value === "object" && !Array.isArray(value);
}
function toRunRelativePosix(runDir, absolutePath) {
    if (!absolutePath)
        return undefined;
    return path.relative(runDir, absolutePath).replace(/\\/g, "/");
}
function normalizeArtifactRef(runDir, ref) {
    const absolute = resolveArtifactAbsolutePath(runDir, ref);
    if (!absolute)
        return null;
    const relative = toRunRelativePosix(runDir, absolute);
    return relative ?? null;
}
function resolveArtifactAbsolutePath(runDir, ref) {
    if (!ref)
        return null;
    const normalized = ref.trim();
    if (!normalized)
        return null;
    const absoluteRunDir = path.resolve(runDir);
    if (path.isAbsolute(normalized) || /^[A-Za-z]:[\\/]/.test(normalized)) {
        return path.normalize(normalized);
    }
    const candidates = collectArtifactCandidates(absoluteRunDir, normalized);
    if (candidates.length > 0) {
        candidates.sort((a, b) => {
            if (a.outsideRun !== b.outsideRun)
                return a.outsideRun ? 1 : -1;
            return a.relative.length - b.relative.length;
        });
        return candidates[0].absolute;
    }
    return path.join(absoluteRunDir, normalized);
}
function collectArtifactCandidates(runDir, ref) {
    const seen = new Map();
    const pushCandidate = (absolute) => {
        const normalizedAbs = path.normalize(absolute);
        const relative = path.relative(runDir, normalizedAbs).replace(/\\/g, "/");
        const outsideRun = relative.startsWith("..");
        seen.set(normalizedAbs, { absolute: normalizedAbs, relative, outsideRun });
    };
    pushCandidate(path.join(runDir, ref));
    pushCandidate(path.resolve(ref));
    return Array.from(seen.values());
}
function defaultResultRef(effectId) {
    return `tasks/${effectId}/result.json`;
}
function formatEntrypointSpecifier(entrypoint) {
    return `${entrypoint.importPath}#${entrypoint.exportName}`;
}
function parseEntrypointSpecifier(specifier) {
    if (!specifier) {
        throw new Error("Entrypoint must be provided as <path>#<export>");
    }
    const hashIndex = specifier.lastIndexOf("#");
    if (hashIndex === 0) {
        throw new Error("Entrypoint must include a module path before '#'");
    }
    if (hashIndex === -1) {
        return { importPath: specifier };
    }
    const importPath = specifier.slice(0, hashIndex);
    if (!importPath) {
        throw new Error("Entrypoint must include a module path before '#'");
    }
    const exportName = specifier.slice(hashIndex + 1) || undefined;
    return { importPath, exportName };
}
// Use an indirect dynamic import so TypeScript does not downlevel to require() in CommonJS builds.
// Vitest executes modules inside a VM context that requires direct import() support.
const dynamicImportModule = (() => {
    if (process.env.VITEST) {
        return (specifier) => Promise.resolve(`${specifier}`).then(s => __importStar(require(s)));
    }
    // eslint-disable-next-line @typescript-eslint/no-implied-eval
    return new Function("specifier", "return import(specifier);");
})();
function listModuleExports(mod) {
    const keys = Object.keys(mod);
    return keys.length > 0 ? keys.join(", ") : "(none)";
}
async function validateProcessEntrypoint(importPath, exportName) {
    const resolvedPath = path.isAbsolute(importPath) ? importPath : (0, resolveInputPath_1.resolveInputPath)(importPath);
    try {
        await node_fs_1.promises.access(resolvedPath);
    }
    catch (error) {
        throw new Error(`Process entry file not found: ${resolvedPath}. ` +
            `Ensure the path is correct and points to a valid JS/TS module.`);
    }
    const moduleUrl = (0, node_url_1.pathToFileURL)(resolvedPath).href;
    let mod;
    try {
        mod = await dynamicImportModule(moduleUrl);
    }
    catch (error) {
        throw new Error(`Failed to load process module at ${resolvedPath}: ${error instanceof Error ? error.message : String(error)}`);
    }
    const resolvedExportName = exportName ?? "process";
    const candidate = (resolvedExportName && mod[resolvedExportName]) ??
        mod.process ??
        mod.default;
    if (typeof candidate !== "function") {
        const available = listModuleExports(mod);
        if (resolvedExportName && !(resolvedExportName in mod) && mod.default) {
            throw new Error(`Process module ${resolvedPath} does not export '${resolvedExportName}'. ` +
                `Available exports: ${available}. ` +
                `If you intended a default export, pass --entry ${importPath}#default.`);
        }
        throw new Error(`Process module ${resolvedPath} does not export a valid process function. ` +
            `Expected '${resolvedExportName}' (function) or default export. ` +
            `Available exports: ${available}.`);
    }
}
async function readInputsFile(filePath) {
    const absolute = (0, resolveInputPath_1.resolveInputPath)(filePath);
    let contents;
    try {
        contents = await node_fs_1.promises.readFile(absolute, "utf8");
    }
    catch (error) {
        throw new Error(`Failed to read inputs file ${absolute}: ${error instanceof Error ? error.message : String(error)}`);
    }
    try {
        return JSON.parse(contents);
    }
    catch (error) {
        throw new Error(`Failed to parse inputs file ${absolute} as JSON: ${error instanceof Error ? error.message : String(error)}`);
    }
}
async function handleRunCreate(parsed) {
    if (!parsed.processId) {
        console.error("--process-id is required for run:create");
        console.error(USAGE);
        return 1;
    }
    if (!parsed.entrySpecifier) {
        console.error("--entry is required for run:create");
        console.error(USAGE);
        return 1;
    }
    let entrypoint;
    try {
        entrypoint = parseEntrypointSpecifier(parsed.entrySpecifier);
    }
    catch (error) {
        console.error(error instanceof Error ? error.message : String(error));
        return 1;
    }
    const runsDir = collapseDoubledA5cRuns(path.resolve(parsed.runsDir));
    const absoluteImportPath = path.resolve(entrypoint.importPath);
    const resolvedEntry = formatResolvedEntrypoint(absoluteImportPath, entrypoint.exportName);
    logVerbose("run:create", parsed, {
        runsDir,
        processId: parsed.processId,
        entry: resolvedEntry,
        dryRun: parsed.dryRun,
        json: parsed.json,
        request: parsed.requestId,
        prompt: parsed.prompt,
        processRevision: parsed.processRevision,
        runId: parsed.runIdOverride,
        inputsPath: parsed.inputsPath ? path.resolve(parsed.inputsPath) : undefined,
    });
    let inputs = undefined;
    if (parsed.inputsPath) {
        try {
            inputs = await readInputsFile(parsed.inputsPath);
        }
        catch (error) {
            console.error(error instanceof Error ? error.message : String(error));
            return 1;
        }
    }
    if (parsed.dryRun) {
        const summary = {
            dryRun: true,
            runsDir,
            processId: parsed.processId,
            entry: resolvedEntry,
            runId: parsed.runIdOverride ?? null,
            request: parsed.requestId ?? null,
            processRevision: parsed.processRevision ?? null,
            inputsPath: parsed.inputsPath ? path.resolve(parsed.inputsPath) : null,
        };
        if (parsed.json) {
            console.log(JSON.stringify(summary, null, 2));
        }
        else {
            const parts = [
                "[run:create] dry-run",
                `runsDir=${runsDir}`,
                `processId=${parsed.processId}`,
                `entry=${resolvedEntry}`,
                `runId=${summary.runId ?? "auto"}`,
            ];
            if (parsed.requestId)
                parts.push(`request=${parsed.requestId}`);
            if (parsed.processRevision)
                parts.push(`processRevision=${parsed.processRevision}`);
            if (summary.inputsPath)
                parts.push(`inputs=${summary.inputsPath}`);
            console.log(parts.join(" "));
        }
        return 0;
    }
    try {
        await validateProcessEntrypoint(absoluteImportPath, entrypoint.exportName);
    }
    catch (error) {
        console.error(error instanceof Error ? error.message : String(error));
        return 1;
    }
    const result = await (0, createRun_1.createRun)({
        runsDir,
        runId: parsed.runIdOverride,
        request: parsed.requestId,
        prompt: parsed.prompt,
        processRevision: parsed.processRevision,
        process: {
            processId: parsed.processId,
            importPath: absoluteImportPath,
            exportName: entrypoint.exportName,
        },
        inputs,
    });
    const entrySpec = formatEntrypointSpecifier(result.metadata.entrypoint);
    // --- Harness-specific session binding ---
    // When --harness is explicitly specified, use that adapter directly.
    // Otherwise auto-detect from env vars (CLAUDE_SESSION_ID, CLAUDE_ENV_FILE).
    const adapter = parsed.harness
        ? (0, harness_1.getAdapterByName)(parsed.harness)
        : (0, harness_1.getAdapter)();
    let sessionBound;
    if (adapter) {
        const sessionId = adapter.resolveSessionId(parsed);
        if (sessionId) {
            sessionBound = await adapter.bindSession({
                sessionId,
                runId: result.runId,
                runDir: result.runDir,
                pluginRoot: adapter.resolvePluginRoot(parsed),
                stateDir: parsed.stateDir,
                maxIterations: parsed.maxIterations,
                prompt: parsed.prompt ?? "",
                verbose: parsed.verbose,
                json: parsed.json,
            });
        }
        else if (parsed.harness) {
            // --harness was specified but no session ID could be resolved
            sessionBound = {
                harness: parsed.harness,
                sessionId: "",
                error: "No session ID provided. Use --session-id or set CLAUDE_SESSION_ID.",
            };
        }
    }
    else if (parsed.harness) {
        // --harness was specified but the adapter name is unknown
        sessionBound = {
            harness: parsed.harness,
            sessionId: "",
            error: `Unsupported harness: ${parsed.harness}`,
        };
    }
    // Discover available skills and agents for the new run
    // Try process-driven discovery first (reads @skill/@agent markers from process file)
    let discoveredSkills;
    let discoveredAgents;
    const discoverPluginRoot = adapter?.resolvePluginRoot(parsed) ?? (0, harness_1.getAdapter)().resolvePluginRoot(parsed);
    if (discoverPluginRoot) {
        try {
            const processDiscovery = (0, skill_1.discoverFromProcessFile)({
                processFilePath: absoluteImportPath,
                pluginRoot: discoverPluginRoot,
            });
            if (processDiscovery) {
                discoveredSkills = processDiscovery.skills;
                discoveredAgents = processDiscovery.agents;
            }
            else {
                // Fallback to generic scan
                const discoverResult = await (0, skill_1.discoverSkillsInternal)({
                    pluginRoot: discoverPluginRoot,
                    runId: result.runId,
                    runsDir: parsed.runsDir,
                    processPath: absoluteImportPath,
                });
                discoveredSkills = discoverResult.skills.map(s => ({ name: s.name, file: s.file }));
                discoveredAgents = discoverResult.agents.map(a => ({ name: a.name, file: a.file }));
            }
        }
        catch {
            // Non-fatal
        }
    }
    if (parsed.json) {
        const compactSkills = discoveredSkills
            ? { count: discoveredSkills.length, names: discoveredSkills.map(s => s.name) }
            : undefined;
        const compactAgents = discoveredAgents
            ? { count: discoveredAgents.length, names: discoveredAgents.map(a => a.name) }
            : undefined;
        console.log(JSON.stringify({
            runId: result.runId,
            runDir: result.runDir,
            entry: entrySpec,
            session: sessionBound ?? undefined,
            discoveredSkills: parsed.verbose ? discoveredSkills : compactSkills,
            discoveredAgents: parsed.verbose ? discoveredAgents : compactAgents,
        }, null, 2));
    }
    else {
        console.log(`[run:create] runId=${result.runId} runDir=${result.runDir} entry=${entrySpec}`);
        if (sessionBound?.error) {
            console.error(`[run:create] Session binding error: ${sessionBound.error}`);
        }
        else if (sessionBound) {
            console.log(`[run:create] session=${sessionBound.sessionId} bound via ${sessionBound.harness} stateFile=${sessionBound.stateFile}`);
        }
    }
    return 0;
}
async function handleRunStatus(parsed) {
    if (!parsed.runDirArg) {
        console.error(USAGE);
        return 1;
    }
    const runDir = resolveRunDir(parsed.runsDir, parsed.runDirArg);
    logVerbose("run:status", parsed, {
        runDir,
        json: parsed.json,
    });
    const metadata = await readRunMetadataSafe(runDir, "run:status");
    if (!metadata)
        return 1;
    const journal = await loadJournalSafe(runDir, "run:status");
    if (!journal)
        return 1;
    const index = await buildEffectIndexSafe(runDir, "run:status", journal);
    if (!index)
        return 1;
    const pendingRecords = index.listPendingEffects();
    const pendingByKind = countPendingByKind(pendingRecords);
    const pendingTotal = pendingRecords.length;
    const stateSnapshot = await readStateCacheSafe(runDir, "run:status");
    const mergedMetadata = mergeMetadataSources({
        pendingEffectsByKind: pendingByKind,
    }, { snapshot: stateSnapshot, pendingByKind });
    const formattedMetadata = formatIterationMetadata(mergedMetadata);
    const lastEvent = journal.at(-1);
    const lastLifecycleEvent = findLastLifecycleEvent(journal);
    const state = deriveRunState(lastLifecycleEvent?.type, pendingTotal);
    const lastSummary = formatLastEventSummary(lastEvent);
    const autoRunnableCount = pendingRecords.filter(r => r.kind === "node").length;
    const pendingEffectsSummary = {
        totalPending: pendingTotal,
        countsByKind: pendingByKind,
        autoRunnableCount,
    };
    const needsMoreIterations = state === "waiting" && autoRunnableCount > 0;
    if (parsed.json) {
        const completionProof = state === "completed" ? (0, completionProof_1.resolveCompletionProof)(metadata) : null;
        console.log(JSON.stringify({
            state,
            lastEvent: lastEvent ? serializeJournalEvent(lastEvent, runDir) : null,
            pendingByKind,
            pendingEffectsSummary,
            needsMoreIterations,
            metadata: formattedMetadata.jsonMetadata ?? null,
            completionProof,
        }));
        return 0;
    }
    const suffix = formattedMetadata.textParts.length ? ` ${formattedMetadata.textParts.join(" ")}` : "";
    const completionProof = state === "completed" ? (0, completionProof_1.resolveCompletionProof)(metadata) : undefined;
    const secretSuffix = completionProof ? ` completionProof=${completionProof}` : "";
    console.log(`[run:status] state=${state} last=${lastSummary}${suffix}${secretSuffix}`);
    return 0;
}
async function handleRunIterate(parsed) {
    if (!parsed.runDirArg) {
        console.error(USAGE);
        return 1;
    }
    const runDir = resolveRunDir(parsed.runsDir, parsed.runDirArg);
    logVerbose("run:iterate", parsed, {
        runDir,
        iteration: parsed.iteration,
        json: parsed.json,
        verbose: parsed.verbose,
    });
    try {
        const result = await (0, runIterate_1.runIterate)({
            runDir,
            iteration: parsed.iteration,
            verbose: parsed.verbose,
            json: parsed.json,
        });
        if (parsed.json) {
            console.log(JSON.stringify(result, null, 2));
        }
        else {
            const countInfo = result.count ? ` count=${result.count}` : "";
            const actionInfo = result.action ? ` action=${result.action}` : "";
            const progressInfo = result.iterationCount > 0 ? ` (${result.iterationCount} completed)` : "";
            console.log(`[run:iterate] iteration=${result.iteration}${progressInfo} status=${result.status}${actionInfo}${countInfo} reason=${result.reason}`);
            if (result.status === "completed" && result.completionProof) {
                console.log(`[run:iterate] completionProof=${result.completionProof}`);
            }
            if (result.status === "waiting" && result.until) {
                console.log(`[run:iterate] Waiting until: ${new Date(result.until).toISOString()}`);
            }
        }
        return 0;
    }
    catch (error) {
        console.error(`[run:iterate] Error: ${error instanceof Error ? error.message : String(error)}`);
        return 1;
    }
}
async function handleRunExecuteTasks(parsed) {
    if (!parsed.runDirArg) {
        console.error(USAGE);
        return 1;
    }
    const runDir = resolveRunDir(parsed.runsDir, parsed.runDirArg);
    logVerbose("run:execute-tasks", parsed, {
        runDir,
        maxTasks: parsed.maxTasks,
        kind: parsed.kindFilter,
        timeout: parsed.timeout,
        dryRun: parsed.dryRun,
        json: parsed.json,
        verbose: parsed.verbose,
    });
    try {
        const result = await (0, runExecuteTasks_1.runExecuteTasks)({
            runDir,
            maxTasks: parsed.maxTasks,
            kind: parsed.kindFilter,
            timeout: parsed.timeout,
            verbose: parsed.verbose,
            json: parsed.json,
            dryRun: parsed.dryRun,
        });
        if (parsed.json) {
            console.log(JSON.stringify(result, null, 2));
        }
        else {
            const taskSummary = result.tasks.length > 0
                ? ` tasks=[${result.tasks.map((t) => `${t.effectId}:${t.status}`).join(",")}]`
                : "";
            console.log(`[run:execute-tasks] action=${result.action} count=${result.count} reason=${result.reason}${taskSummary}`);
        }
        return 0;
    }
    catch (error) {
        console.error(`[run:execute-tasks] Error: ${error instanceof Error ? error.message : String(error)}`);
        return 1;
    }
}
async function handleRunEvents(parsed) {
    if (!parsed.runDirArg) {
        console.error(USAGE);
        return 1;
    }
    const runDir = resolveRunDir(parsed.runsDir, parsed.runDirArg);
    logVerbose("run:events", parsed, {
        runDir,
        json: parsed.json,
        limit: parsed.limit,
        reverse: parsed.reverseOrder,
        filterType: parsed.filterType,
    });
    if (!(await readRunMetadataSafe(runDir, "run:events")))
        return 1;
    const stateSnapshot = await readStateCacheSafe(runDir, "run:events");
    const journal = await loadJournalSafe(runDir, "run:events");
    if (!journal)
        return 1;
    const filterType = parsed.filterType ? parsed.filterType.toUpperCase() : undefined;
    const filtered = filterType ? journal.filter((event) => event.type.toUpperCase() === filterType) : journal;
    const orderedBase = filtered.slice();
    const ordered = parsed.reverseOrder ? orderedBase.reverse() : orderedBase;
    const limited = parsed.limit !== undefined ? ordered.slice(0, parsed.limit) : ordered;
    const metadata = mergeMetadataSources(undefined, { snapshot: stateSnapshot });
    const formattedMetadata = formatIterationMetadata(metadata);
    if (parsed.json) {
        console.log(JSON.stringify({
            events: limited.map((event) => serializeJournalEvent(event, runDir)),
            metadata: formattedMetadata.jsonMetadata ?? null,
        }));
        return 0;
    }
    const headerParts = [
        `total=${journal.length}`,
        `matching=${filtered.length}`,
        `showing=${limited.length}`,
    ];
    if (filterType)
        headerParts.push(`filter=${filterType}`);
    if (parsed.limit)
        headerParts.push(`limit=${parsed.limit}`);
    if (parsed.reverseOrder)
        headerParts.push("order=desc");
    const metadataSuffix = formattedMetadata.textParts.length ? ` ${formattedMetadata.textParts.join(" ")}` : "";
    console.log(`[run:events] ${headerParts.join(" ")}${metadataSuffix}`);
    for (const event of limited) {
        console.log(`- ${formatEventLine(event)}`);
    }
    return 0;
}
async function handleRunRebuildState(parsed) {
    if (!parsed.runDirArg) {
        console.error(USAGE);
        return 1;
    }
    const runDir = resolveRunDir(parsed.runsDir, parsed.runDirArg);
    logVerbose("run:rebuild-state", parsed, {
        runDir,
        dryRun: parsed.dryRun,
        json: parsed.json,
    });
    if (!(await readRunMetadataSafe(runDir, "run:rebuild-state")))
        return 1;
    if (parsed.dryRun) {
        const plan = { dryRun: true, runDir, plan: "rebuild_state_cache", reason: "cli_manual" };
        if (parsed.json) {
            console.log(JSON.stringify(plan, null, 2));
        }
        else {
            console.log(`[run:rebuild-state] dry-run runDir=${runDir} plan=${plan.plan} reason=${plan.reason}`);
        }
        return 0;
    }
    const snapshot = await (0, stateCache_1.rebuildStateCache)(runDir, { reason: "cli_manual" });
    const metadata = {
        pendingEffectsByKind: snapshot.pendingEffectsByKind,
        stateVersion: snapshot.stateVersion,
        journalHead: snapshot.journalHead ?? null,
        stateRebuilt: true,
        stateRebuildReason: snapshot.rebuildReason ?? undefined,
    };
    const formatted = formatIterationMetadata(metadata);
    if (parsed.json) {
        console.log(JSON.stringify({ runDir, metadata: formatted.jsonMetadata ?? null }, null, 2));
        return 0;
    }
    const suffix = formatted.textParts.length ? ` ${formatted.textParts.join(" ")}` : "";
    console.log(`[run:rebuild-state] runDir=${runDir}${suffix}`);
    return 0;
}
async function handleRunRepairJournal(parsed) {
    if (!parsed.runDirArg) {
        console.error(USAGE);
        return 1;
    }
    const runDir = resolveRunDir(parsed.runsDir, parsed.runDirArg);
    logVerbose("run:repair-journal", parsed, {
        runDir,
        dryRun: parsed.dryRun,
        json: parsed.json,
    });
    if (!(await readRunMetadataSafe(runDir, "run:repair-journal")))
        return 1;
    const journalDir = path.join(runDir, "journal");
    const files = (await node_fs_1.promises.readdir(journalDir)).filter((name) => name.endsWith(".json")).sort();
    const rawEvents = [];
    for (const filename of files) {
        const fullPath = path.join(journalDir, filename);
        const payload = JSON.parse(await node_fs_1.promises.readFile(fullPath, "utf8"));
        rawEvents.push({ filename, payload });
    }
    const seenInvocation = new Set();
    const keptEffectIds = new Set();
    const droppedEffectIds = new Set();
    const kept = [];
    let droppedRequested = 0;
    let droppedResolved = 0;
    for (const entry of rawEvents) {
        const type = typeof entry.payload.type === "string" ? entry.payload.type : "UNKNOWN";
        const recordedAt = typeof entry.payload.recordedAt === "string" ? entry.payload.recordedAt : undefined;
        const data = isJsonRecord(entry.payload.data) ? entry.payload.data : {};
        if (type === "EFFECT_REQUESTED") {
            const invocationKey = typeof data.invocationKey === "string" ? data.invocationKey : "";
            const effectId = typeof data.effectId === "string" ? data.effectId : "";
            if (invocationKey && seenInvocation.has(invocationKey)) {
                droppedRequested += 1;
                if (effectId)
                    droppedEffectIds.add(effectId);
                continue;
            }
            if (invocationKey)
                seenInvocation.add(invocationKey);
            if (effectId)
                keptEffectIds.add(effectId);
            kept.push({ type, recordedAt, data });
            continue;
        }
        if (type === "EFFECT_RESOLVED") {
            const effectId = typeof data.effectId === "string" ? data.effectId : "";
            if (effectId && droppedEffectIds.has(effectId) && !keptEffectIds.has(effectId)) {
                droppedResolved += 1;
                continue;
            }
            kept.push({ type, recordedAt, data });
            continue;
        }
        // Keep all other events.
        kept.push({ type, recordedAt, data });
    }
    const summary = {
        runDir,
        journal: {
            originalFiles: files.length,
            keptEvents: kept.length,
            droppedRequested,
            droppedResolved,
        },
    };
    if (parsed.dryRun) {
        if (parsed.json) {
            console.log(JSON.stringify({ dryRun: true, ...summary }, null, 2));
        }
        else {
            console.log(`[run:repair-journal] dry-run originalFiles=${files.length} keptEvents=${kept.length} droppedRequested=${droppedRequested} droppedResolved=${droppedResolved}`);
        }
        return 0;
    }
    const stamp = Date.now();
    const repairedDir = path.join(runDir, `journal.repaired.${stamp}`);
    await node_fs_1.promises.mkdir(repairedDir, { recursive: true });
    for (let i = 0; i < kept.length; i += 1) {
        const seq = String(i + 1).padStart(6, "0");
        const ulid = (0, ulids_1.nextUlid)();
        const filename = `${seq}.${ulid}.json`;
        const eventPayload = {
            type: kept[i].type,
            recordedAt: kept[i].recordedAt ?? new Date().toISOString(),
            data: kept[i].data,
        };
        const contents = JSON.stringify(eventPayload, null, 2) + "\n";
        const checksum = crypto.createHash("sha256").update(contents).digest("hex");
        const withChecksum = JSON.stringify({ ...eventPayload, checksum }, null, 2) + "\n";
        await node_fs_1.promises.writeFile(path.join(repairedDir, filename), withChecksum, "utf8");
    }
    const backupDir = path.join(runDir, `journal.bak.${stamp}`);
    await node_fs_1.promises.rename(journalDir, backupDir);
    await node_fs_1.promises.rename(repairedDir, journalDir);
    if (parsed.json) {
        console.log(JSON.stringify({ ...summary, backupDir, repaired: true }, null, 2));
    }
    else {
        console.log(`[run:repair-journal] repaired originalFiles=${files.length} keptEvents=${kept.length} droppedRequested=${droppedRequested} droppedResolved=${droppedResolved} backupDir=${backupDir}`);
    }
    return 0;
}
async function handleTaskPost(parsed) {
    if (!parsed.runDirArg || !parsed.effectId) {
        console.error(USAGE);
        return 1;
    }
    if (!parsed.taskStatus) {
        console.error(`[task:post] missing required --status <ok|error>`);
        return 1;
    }
    if (parsed.stdoutRef && parsed.stdoutFile) {
        console.error(`[task:post] cannot combine --stdout-ref with --stdout-file`);
        return 1;
    }
    if (parsed.stderrRef && parsed.stderrFile) {
        console.error(`[task:post] cannot combine --stderr-ref with --stderr-file`);
        return 1;
    }
    const runDir = resolveRunDir(parsed.runsDir, parsed.runDirArg);
    const secretLogsAllowed = allowSecretLogs(parsed);
    logVerbose("task:post", parsed, {
        runDir,
        effectId: parsed.effectId,
        status: parsed.taskStatus,
        dryRun: parsed.dryRun,
        json: parsed.json,
        secretLogsAllowed,
    });
    const index = await buildEffectIndexSafe(runDir, "task:post");
    if (!index)
        return 1;
    const record = index.getByEffectId(parsed.effectId);
    if (!record) {
        console.error(`[task:post] effect ${parsed.effectId} not found at ${runDir}`);
        return 1;
    }
    if (record.status !== "requested") {
        console.error(`[task:post] effect ${parsed.effectId} is not requested (status=${record.status ?? "unknown"})`);
        return 1;
    }
    const nowIso = new Date().toISOString();
    const startedAt = parsed.startedAt ?? nowIso;
    const finishedAt = parsed.finishedAt ?? nowIso;
    const resolveMaybeRunRelative = (candidate) => {
        if (!candidate)
            return undefined;
        if (candidate === "-")
            return candidate;
        if (path.isAbsolute(candidate) || /^[A-Za-z]:[\\/]/.test(candidate)) {
            return candidate;
        }
        return path.join(runDir, candidate);
    };
    const readJsonFile = async (_label, filename) => {
        if (!filename)
            return undefined;
        if (filename === "-") {
            const raw = await readStdinUtf8();
            const trimmed = raw.trim();
            return trimmed.length ? JSON.parse(trimmed) : undefined;
        }
        const absolute = resolveMaybeRunRelative(filename);
        const raw = await node_fs_1.promises.readFile(absolute, "utf8");
        const trimmed = raw.trim();
        return trimmed.length ? JSON.parse(trimmed) : undefined;
    };
    const readTextFile = async (_label, filename) => {
        if (!filename)
            return undefined;
        if (filename === "-") {
            return await readStdinUtf8();
        }
        const absolute = resolveMaybeRunRelative(filename);
        return await node_fs_1.promises.readFile(absolute, "utf8");
    };
    const metadataRaw = await readJsonFile("metadata", parsed.metadataPath);
    const metadata = isJsonRecord(metadataRaw) ? metadataRaw : undefined;
    const stdout = parsed.stdoutFile ? await readTextFile("stdout", parsed.stdoutFile) : undefined;
    const stderr = parsed.stderrFile ? await readTextFile("stderr", parsed.stderrFile) : undefined;
    let value = undefined;
    let errorPayload = undefined;
    if (parsed.taskStatus === "ok") {
        value = await readJsonFile("value", parsed.valuePath);
    }
    else {
        errorPayload =
            (await readJsonFile("error", parsed.errorPath)) ??
                {
                    name: "Error",
                    message: "Task reported failure",
                };
    }
    const invocationKey = parsed.invocationKey ?? record.invocationKey;
    const plan = {
        runDir: toRunRelativePosix(runDir, runDir) ?? runDir,
        effectId: parsed.effectId,
        status: parsed.taskStatus,
        valueProvided: parsed.valuePath ? true : false,
        errorProvided: parsed.errorPath ? true : false,
        stdoutRef: parsed.stdoutRef ?? null,
        stderrRef: parsed.stderrRef ?? null,
        stdoutFile: parsed.stdoutFile ?? null,
        stderrFile: parsed.stderrFile ?? null,
    };
    if (parsed.dryRun) {
        if (parsed.json) {
            console.log(JSON.stringify({ status: "skipped", dryRun: true, plan }, null, 2));
        }
        else {
            console.log(`[task:post] status=skipped`);
            console.error(`[task:post] dry-run plan ${JSON.stringify(plan)}`);
        }
        return 0;
    }
    const committed = await (0, commitEffectResult_1.commitEffectResult)({
        runDir,
        effectId: parsed.effectId,
        invocationKey,
        result: parsed.taskStatus === "ok"
            ? {
                status: "ok",
                value,
                stdout,
                stderr,
                stdoutRef: parsed.stdoutRef,
                stderrRef: parsed.stderrRef,
                startedAt,
                finishedAt,
                metadata,
            }
            : {
                status: "error",
                error: errorPayload,
                stdout,
                stderr,
                stdoutRef: parsed.stdoutRef,
                stderrRef: parsed.stderrRef,
                startedAt,
                finishedAt,
                metadata,
            },
    });
    const stdoutRef = normalizeArtifactRef(runDir, committed.stdoutRef) ?? null;
    const stderrRef = normalizeArtifactRef(runDir, committed.stderrRef) ?? null;
    const resultRef = normalizeArtifactRef(runDir, committed.resultRef) ?? null;
    if (parsed.json) {
        console.log(JSON.stringify({
            status: parsed.taskStatus,
            committed,
            stdoutRef,
            stderrRef,
            resultRef,
        }));
    }
    else {
        const parts = [`[task:post] status=${parsed.taskStatus}`];
        if (stdoutRef)
            parts.push(`stdoutRef=${stdoutRef}`);
        if (stderrRef)
            parts.push(`stderrRef=${stderrRef}`);
        if (resultRef)
            parts.push(`resultRef=${resultRef}`);
        console.log(parts.join(" "));
    }
    return parsed.taskStatus === "ok" ? 0 : 1;
}
async function handleTaskList(parsed) {
    if (!parsed.runDirArg) {
        console.error(USAGE);
        return 1;
    }
    const runDir = resolveRunDir(parsed.runsDir, parsed.runDirArg);
    logVerbose("task:list", parsed, {
        runDir,
        json: parsed.json,
        pending: parsed.pendingOnly,
        kind: parsed.kindFilter,
    });
    const index = await buildEffectIndexSafe(runDir, "task:list");
    if (!index)
        return 1;
    const rawRecords = parsed.pendingOnly ? index.listPendingEffects() : index.listEffects();
    const records = rawRecords
        .filter((record) => parsed.kindFilter ? record.kind?.toLowerCase() === parsed.kindFilter.toLowerCase() : true)
        .sort((a, b) => a.effectId.localeCompare(b.effectId));
    const entries = records.map((record) => toTaskListEntry(record, runDir));
    if (parsed.json) {
        console.log(JSON.stringify({ tasks: entries }, null, 2));
        return 0;
    }
    const scope = parsed.pendingOnly ? "pending" : "total";
    console.log(`[task:list] ${scope}=${entries.length}`);
    for (const entry of entries) {
        const label = entry.label ? ` ${entry.label}` : "";
        console.log(`- ${entry.effectId} [${entry.kind ?? "unknown"} ${entry.status}]${label} (taskId=${entry.taskId ?? "n/a"})`);
    }
    return 0;
}
async function handleTaskShow(parsed) {
    if (!parsed.runDirArg || !parsed.effectId) {
        console.error(USAGE);
        return 1;
    }
    const runDir = resolveRunDir(parsed.runsDir, parsed.runDirArg);
    const secretLogsAllowed = allowSecretLogs(parsed);
    logVerbose("task:show", parsed, {
        runDir,
        effectId: parsed.effectId,
        json: parsed.json,
        secretLogsAllowed,
    });
    const index = await buildEffectIndexSafe(runDir, "task:show");
    if (!index)
        return 1;
    const record = index.getByEffectId(parsed.effectId);
    if (!record) {
        console.error(`[task:show] effect ${parsed.effectId} not found in ${runDir}`);
        return 1;
    }
    const taskDef = await (0, tasks_1.readTaskDefinition)(runDir, parsed.effectId);
    if (!taskDef) {
        console.error(`[task:show] task definition missing for effect ${parsed.effectId}`);
        return 1;
    }
    const preview = await loadTaskResultPreview(runDir, parsed.effectId, record);
    const entry = toTaskListEntry(record, runDir);
    const inlineResult = preview.large ? null : preview.result ?? null;
    const largeResultRef = preview.large ? entry.resultRef ?? defaultResultRef(record.effectId) : null;
    if (parsed.json) {
        console.log(JSON.stringify({
            effect: entry,
            task: secretLogsAllowed ? taskDef : null,
            result: secretLogsAllowed ? inlineResult : null,
            largeResult: largeResultRef,
        }));
        return 0;
    }
    console.log(`[task:show] ${entry.effectId} [${entry.kind ?? "unknown"} ${entry.status}] ${entry.label ?? "(no label)"} (taskId=${entry.taskId})`);
    console.log(`  stepId=${entry.stepId} requestedAt=${entry.requestedAt ?? "n/a"} resolvedAt=${entry.resolvedAt ?? "n/a"}`);
    console.log(`  taskDefRef=${entry.taskDefRef ?? "n/a"}`);
    console.log(`  inputsRef=${entry.inputsRef ?? "n/a"}`);
    console.log(`  resultRef=${entry.resultRef ?? "n/a"}`);
    console.log(`  stdoutRef=${entry.stdoutRef ?? "n/a"}`);
    console.log(`  stderrRef=${entry.stderrRef ?? "n/a"}`);
    if (!secretLogsAllowed) {
        console.log("  payloads: redacted (set BABYSITTER_ALLOW_SECRET_LOGS=true and rerun with --json --verbose to view task/result blobs)");
        if (!inlineResult && !preview.large) {
            console.log("  result: (not yet written)");
        }
        return 0;
    }
    console.log("  taskDef:", JSON.stringify(taskDef, null, 2));
    if (preview.large) {
        console.log(`  result: see ${largeResultRef ?? entry.resultRef ?? defaultResultRef(record.effectId)}`);
    }
    else if (inlineResult) {
        console.log("  result:", JSON.stringify(inlineResult, null, 2));
    }
    else {
        console.log("  result: (not yet written)");
    }
    return 0;
}
function toTaskListEntry(record, runDir) {
    return {
        effectId: record.effectId,
        taskId: record.taskId ?? "unknown",
        stepId: record.stepId ?? "unknown",
        status: record.status ?? "unknown",
        kind: record.kind,
        label: record.label,
        labels: record.labels,
        taskDefRef: normalizeArtifactRef(runDir, record.taskDefRef ?? `tasks/${record.effectId}/task.json`),
        inputsRef: normalizeArtifactRef(runDir, record.inputsRef),
        resultRef: normalizeArtifactRef(runDir, record.resultRef),
        stdoutRef: normalizeArtifactRef(runDir, record.stdoutRef),
        stderrRef: normalizeArtifactRef(runDir, record.stderrRef),
        requestedAt: record.requestedAt,
        resolvedAt: record.resolvedAt,
    };
}
async function buildEffectIndexSafe(runDir, command, events) {
    try {
        return await (0, effectIndex_1.buildEffectIndex)({ runDir, events });
    }
    catch (error) {
        console.error(`[${command}] unable to read run at ${runDir}: ${error instanceof Error ? error.message : String(error)}`);
        return null;
    }
}
async function readRunMetadataSafe(runDir, command) {
    try {
        return await (0, runFiles_1.readRunMetadata)(runDir);
    }
    catch (error) {
        console.error(`[${command}] unable to read run metadata at ${runDir}: ${error instanceof Error ? error.message : String(error)}`);
        return null;
    }
}
async function loadJournalSafe(runDir, command) {
    try {
        return await (0, journal_1.loadJournal)(runDir);
    }
    catch (error) {
        console.error(`[${command}] unable to read journal at ${runDir}: ${error instanceof Error ? error.message : String(error)}`);
        return null;
    }
}
async function readStateCacheSafe(runDir, command) {
    try {
        const snapshot = await (0, stateCache_1.readStateCache)(runDir);
        if (!snapshot) {
            console.warn(`[${command}] state cache snapshot missing at ${runDir} (continuing without metadata)`);
            return null;
        }
        return snapshot;
    }
    catch (error) {
        console.warn(`[${command}] unable to read state cache at ${runDir}: ${error instanceof Error ? error.message : String(error)} (continuing without metadata)`);
        return null;
    }
}
const RUN_LIFECYCLE_TYPES = new Set([
    "RUN_CREATED",
    "RUN_COMPLETED",
    "RUN_FAILED",
]);
function findLastLifecycleEvent(events) {
    for (let i = events.length - 1; i >= 0; i -= 1) {
        const event = events[i];
        if (RUN_LIFECYCLE_TYPES.has(event.type)) {
            return event;
        }
    }
    return undefined;
}
function countPendingByKind(records) {
    const counts = new Map();
    for (const record of records) {
        const key = record.kind ?? "unknown";
        counts.set(key, (counts.get(key) ?? 0) + 1);
    }
    return Object.fromEntries(Array.from(counts.entries()).sort(([a], [b]) => a.localeCompare(b)));
}
async function loadTaskResultPreview(runDir, effectId, record) {
    const absolutePath = resolveArtifactAbsolutePath(runDir, record.resultRef ?? defaultResultRef(effectId));
    if (!absolutePath)
        return { result: undefined, large: false };
    try {
        const stats = await node_fs_1.promises.stat(absolutePath);
        if (stats.size > LARGE_RESULT_PREVIEW_LIMIT) {
            return { result: undefined, large: true };
        }
    }
    catch (error) {
        const err = error;
        if (err.code === "ENOENT") {
            return { result: undefined, large: false };
        }
        throw error;
    }
    const data = await (0, tasks_1.readTaskResult)(runDir, effectId, record.resultRef);
    return { result: data ?? undefined, large: false };
}
async function readStdinUtf8() {
    return await new Promise((resolve, reject) => {
        const chunks = [];
        process.stdin.on("data", (chunk) => chunks.push(Buffer.from(chunk)));
        process.stdin.on("end", () => resolve(Buffer.concat(chunks).toString("utf8")));
        process.stdin.on("error", reject);
    });
}
function deriveRunState(lastLifecycleEventType, pendingTotal) {
    if (lastLifecycleEventType === "RUN_COMPLETED")
        return "completed";
    if (lastLifecycleEventType === "RUN_FAILED")
        return "failed";
    if (pendingTotal > 0)
        return "waiting";
    return "created";
}
function formatLastEventSummary(event) {
    if (!event)
        return "none";
    return `${event.type}#${formatSeq(event.seq)} ${event.recordedAt}`;
}
function mergeMetadataSources(metadata, options) {
    const snapshot = options.snapshot ?? null;
    const hasPendingOverride = options.pendingByKind !== undefined;
    const snapshotHasInfo = Boolean(snapshot);
    if (!metadata && !hasPendingOverride && !snapshotHasInfo) {
        return undefined;
    }
    const next = { ...(metadata ?? {}) };
    if (hasPendingOverride) {
        next.pendingEffectsByKind = { ...(options.pendingByKind ?? {}) };
    }
    else if (!next.pendingEffectsByKind && snapshot) {
        next.pendingEffectsByKind = { ...snapshot.pendingEffectsByKind };
    }
    if (snapshot) {
        if (next.stateVersion === undefined) {
            next.stateVersion = snapshot.stateVersion;
        }
        if (next.journalHead === undefined) {
            next.journalHead = snapshot.journalHead ?? null;
        }
        if (snapshot.rebuildReason) {
            next.stateRebuilt = true;
            if (!next.stateRebuildReason) {
                next.stateRebuildReason = snapshot.rebuildReason;
            }
        }
    }
    if (next.stateVersion === undefined &&
        next.stateRebuilt === undefined &&
        next.pendingEffectsByKind === undefined &&
        next.journalHead === undefined) {
        return undefined;
    }
    return next;
}
function formatIterationMetadata(metadata) {
    const textParts = [];
    if (!metadata) {
        return { textParts, jsonMetadata: undefined };
    }
    if (metadata.stateVersion !== undefined) {
        textParts.push(`stateVersion=${metadata.stateVersion}`);
    }
    if (metadata.journalHead && typeof metadata.journalHead.seq === "number") {
        const seq = formatSeq(metadata.journalHead.seq);
        textParts.push(`journalHead=#${seq}`);
        if (metadata.journalHead.ulid) {
            textParts.push(`journalHead.ulid=${metadata.journalHead.ulid}`);
        }
        if (metadata.journalHead.checksum) {
            textParts.push(`journalHead.checksum=${metadata.journalHead.checksum}`);
        }
    }
    if (metadata.stateRebuilt) {
        const reasonSuffix = metadata.stateRebuildReason ? `(${metadata.stateRebuildReason})` : "";
        textParts.push(`stateRebuilt=true${reasonSuffix}`);
    }
    if (metadata.pendingEffectsByKind) {
        const pendingEntries = Object.entries(metadata.pendingEffectsByKind).sort(([a], [b]) => a.localeCompare(b));
        const pendingTotal = pendingEntries.reduce((sum, [, count]) => sum + count, 0);
        textParts.push(`pending[total]=${pendingTotal}`);
        for (const [kind, count] of pendingEntries) {
            textParts.push(`pending[${kind}]=${count}`);
        }
    }
    return { textParts, jsonMetadata: metadata };
}
function serializeJournalEvent(event, runDir) {
    const data = ensureIterationMetadata(event.data);
    return {
        seq: event.seq,
        ulid: event.ulid,
        type: event.type,
        recordedAt: event.recordedAt,
        filename: event.filename,
        path: toRunRelativePosix(runDir, event.path),
        data,
    };
}
function ensureIterationMetadata(data) {
    const iteration = data.iteration;
    if (!iteration || typeof iteration !== "object" || Array.isArray(iteration)) {
        return data;
    }
    const iterationRecord = iteration;
    const metadata = iterationRecord.metadata;
    return {
        ...data,
        iteration: {
            ...iterationRecord,
            metadata: metadata === undefined ? null : metadata,
        },
    };
}
function formatEventLine(event) {
    return `#${formatSeq(event.seq)} ${event.type} ${event.recordedAt}`;
}
function formatSeq(seq) {
    return seq.toString().padStart(6, "0");
}
async function readCliVersion() {
    const packagePath = path.join(__dirname, "..", "..", "package.json");
    const raw = await node_fs_1.promises.readFile(packagePath, "utf8");
    const parsed = JSON.parse(raw);
    return parsed.version ?? "unknown";
}
/**
 * Checks if stdout/stderr supports colors
 */
function supportsColors() {
    // Check NO_COLOR environment variable (https://no-color.org/)
    if (process.env.NO_COLOR !== undefined) {
        return false;
    }
    // Check FORCE_COLOR environment variable
    if (process.env.FORCE_COLOR !== undefined) {
        return true;
    }
    // Check if running in a TTY
    if (process.stderr && typeof process.stderr.isTTY === "boolean") {
        return process.stderr.isTTY;
    }
    return false;
}
/**
 * Valid CLI commands
 */
const VALID_COMMANDS = [
    "run:create",
    "run:status",
    "run:iterate",
    "run:events",
    "run:rebuild-state",
    "run:repair-journal",
    "run:execute-tasks",
    "task:post",
    "task:list",
    "task:show",
    "session:init",
    "session:associate",
    "session:resume",
    "session:state",
    "session:update",
    "session:check-iteration",
    "session:last-message",
    "session:iteration-message",
    "hook:log",
    "hook:run",
    "skill:discover",
    "skill:fetch-remote",
    "profile:read",
    "profile:write",
    "profile:merge",
    "profile:render",
    "health",
    "configure",
    "version",
];
/**
 * Handles unknown commands with suggestions
 */
function handleUnknownCommand(command, json) {
    const suggestion = (0, exceptions_1.suggestCommand)(command);
    const suggestions = suggestion ? [`Did you mean: ${suggestion}?`] : [];
    const nextSteps = ["Run with --help to see all available commands"];
    const error = new exceptions_1.BabysitterRuntimeError("UnknownCommandError", `Unknown command: ${command}`, {
        category: exceptions_1.ErrorCategory.Validation,
        suggestions,
        nextSteps,
        details: { command, validCommands: VALID_COMMANDS },
    });
    if (json) {
        console.error(JSON.stringify((0, exceptions_1.toStructuredError)(error), null, 2));
    }
    else {
        const colors = supportsColors();
        console.error((0, exceptions_1.formatErrorWithContext)(error, { colors }));
    }
    return 1;
}
/**
 * Shows configuration before executing a command when --show-config is provided
 */
async function showConfigBeforeCommand(parsed) {
    const { configureShow } = await Promise.resolve().then(() => __importStar(require("./commands/configure")));
    const result = configureShow({ json: parsed.json, defaultsOnly: false });
    if (parsed.json) {
        console.error(JSON.stringify({ showConfig: result }, null, 2));
    }
    else {
        console.error("[show-config] Current effective configuration:");
        for (const item of result.values) {
            const sourceTag = item.source === "env" ? " (env)" : "";
            console.error(`  ${item.key}=${formatVerboseValue(item.value)}${sourceTag}`);
        }
        console.error("");
    }
}
/**
 * Formats and outputs an error in the appropriate format
 */
function outputError(error, options) {
    const { json, verbose = false } = options;
    if (json) {
        console.error(JSON.stringify((0, exceptions_1.toStructuredError)(error, verbose)));
    }
    else {
        const colors = supportsColors();
        if ((0, exceptions_1.isBabysitterError)(error)) {
            console.error((0, exceptions_1.formatErrorWithContext)(error, { colors, includeStack: verbose }));
        }
        else {
            // For non-babysitter errors, wrap them with basic formatting
            const wrappedError = new exceptions_1.BabysitterRuntimeError(error.name || "Error", error.message, {
                category: exceptions_1.ErrorCategory.Internal,
                nextSteps: ["If this error persists, please report it as a bug"],
            });
            console.error((0, exceptions_1.formatErrorWithContext)(wrappedError, { colors, includeStack: verbose }));
        }
    }
}
function createBabysitterCli() {
    return {
        async run(argv = process.argv.slice(2)) {
            let parsedJson = false;
            let parsedVerbose = false;
            try {
                const parsed = parseArgs(argv);
                parsedJson = parsed.json;
                parsedVerbose = parsed.verbose;
                if (parsed.command === "version") {
                    console.log(await readCliVersion());
                    return 0;
                }
                if (!parsed.command || parsed.helpRequested) {
                    console.log(USAGE);
                    return 0;
                }
                // Show config if --show-config flag is provided
                if (parsed.showConfig) {
                    await showConfigBeforeCommand(parsed);
                }
                // Check for valid commands and provide suggestions for unknown ones
                if (!VALID_COMMANDS.includes(parsed.command)) {
                    return handleUnknownCommand(parsed.command, parsed.json);
                }
                if (parsed.command === "run:create") {
                    return await handleRunCreate(parsed);
                }
                if (parsed.command === "run:rebuild-state") {
                    return await handleRunRebuildState(parsed);
                }
                if (parsed.command === "run:repair-journal") {
                    return await handleRunRepairJournal(parsed);
                }
                if (parsed.command === "run:status") {
                    return await handleRunStatus(parsed);
                }
                if (parsed.command === "run:iterate") {
                    return await handleRunIterate(parsed);
                }
                if (parsed.command === "run:execute-tasks") {
                    return await handleRunExecuteTasks(parsed);
                }
                if (parsed.command === "run:events") {
                    return await handleRunEvents(parsed);
                }
                if (parsed.command === "task:post") {
                    return await handleTaskPost(parsed);
                }
                if (parsed.command === "task:list") {
                    return await handleTaskList(parsed);
                }
                if (parsed.command === "task:show") {
                    return await handleTaskShow(parsed);
                }
                // Session commands
                if (parsed.command === "session:init") {
                    return await (0, session_1.handleSessionInit)({
                        sessionId: parsed.sessionId,
                        stateDir: parsed.stateDir,
                        maxIterations: parsed.maxIterations,
                        runId: parsed.runIdOverride,
                        prompt: parsed.prompt,
                        json: parsed.json,
                    });
                }
                if (parsed.command === "session:associate") {
                    return await (0, session_1.handleSessionAssociate)({
                        sessionId: parsed.sessionId,
                        stateDir: parsed.stateDir,
                        runId: parsed.runIdOverride,
                        json: parsed.json,
                    });
                }
                if (parsed.command === "session:resume") {
                    return await (0, session_1.handleSessionResume)({
                        sessionId: parsed.sessionId,
                        stateDir: parsed.stateDir,
                        runId: parsed.runIdOverride,
                        maxIterations: parsed.maxIterations,
                        runsDir: parsed.runsDir,
                        json: parsed.json,
                    });
                }
                if (parsed.command === "session:state") {
                    return await (0, session_1.handleSessionState)({
                        sessionId: parsed.sessionId,
                        stateDir: parsed.stateDir,
                        json: parsed.json,
                    });
                }
                if (parsed.command === "session:update") {
                    return await (0, session_1.handleSessionUpdate)({
                        sessionId: parsed.sessionId,
                        stateDir: parsed.stateDir,
                        iteration: parsed.iteration,
                        lastIterationAt: parsed.lastIterationAt,
                        iterationTimes: parsed.iterationTimes,
                        delete: parsed.deleteSession,
                        json: parsed.json,
                    });
                }
                if (parsed.command === "session:check-iteration") {
                    return await (0, session_1.handleSessionCheckIteration)({
                        sessionId: parsed.sessionId,
                        stateDir: parsed.stateDir,
                        json: parsed.json,
                    });
                }
                if (parsed.command === "session:last-message") {
                    if (!parsed.transcriptPath) {
                        console.error("--transcript-path is required for session:last-message");
                        console.error(USAGE);
                        return 1;
                    }
                    return (0, session_1.handleSessionLastMessage)({
                        transcriptPath: parsed.transcriptPath,
                        json: parsed.json,
                    });
                }
                if (parsed.command === "session:iteration-message") {
                    return await (0, session_1.handleSessionIterationMessage)({
                        iteration: parsed.iteration,
                        runId: parsed.runIdOverride,
                        runsDir: parsed.runsDir,
                        pluginRoot: parsed.pluginRoot,
                        json: parsed.json,
                    });
                }
                // Hook commands
                if (parsed.command === "hook:log") {
                    return await (0, hookLog_1.handleHookLog)({
                        hookType: parsed.hookType ?? "",
                        logFile: parsed.logFile ?? "",
                        json: parsed.json,
                    });
                }
                if (parsed.command === "hook:run") {
                    return await (0, hookRun_1.handleHookRun)({
                        hookType: parsed.hookType ?? "",
                        harness: parsed.harness ?? "claude-code",
                        pluginRoot: parsed.pluginRoot,
                        stateDir: parsed.stateDir,
                        runsDir: parsed.runsDir,
                        json: parsed.json,
                        verbose: parsed.verbose,
                    });
                }
                // Skill commands
                if (parsed.command === "skill:discover") {
                    return await (0, skill_1.handleSkillDiscover)({
                        pluginRoot: parsed.pluginRoot,
                        runId: parsed.runIdOverride,
                        cacheTtl: parsed.cacheTtl,
                        runsDir: parsed.runsDir,
                        json: parsed.json,
                        includeRemote: parsed.includeRemote,
                        summaryOnly: parsed.summaryOnly,
                        processPath: parsed.processPath,
                    });
                }
                if (parsed.command === "skill:fetch-remote") {
                    return await (0, skill_1.handleSkillFetchRemote)({
                        sourceType: parsed.sourceType,
                        url: parsed.url,
                        json: parsed.json,
                    });
                }
                // Profile commands
                if (parsed.command === "profile:read" ||
                    parsed.command === "profile:write" ||
                    parsed.command === "profile:merge" ||
                    parsed.command === "profile:render") {
                    const subcommand = parsed.command.split(":")[1];
                    const profileArgs = {
                        subcommand: subcommand,
                        user: parsed.profileUser ?? false,
                        project: parsed.profileProject ?? false,
                        inputPath: parsed.profileInputPath,
                        dir: parsed.profileDir,
                        json: parsed.json,
                    };
                    return await (0, profile_1.handleProfileCommand)(subcommand, profileArgs);
                }
                if (parsed.command === "health") {
                    return await (0, health_1.handleHealthCommand)({
                        json: parsed.json,
                        verbose: parsed.verbose,
                    });
                }
                if (parsed.command === "configure") {
                    const args = parsed.configureSubcommand ? [parsed.configureSubcommand] : [];
                    return await (0, configure_1.handleConfigureCommand)(args, {
                        json: parsed.json,
                        defaultsOnly: parsed.defaultsOnly,
                    });
                }
                // This should not be reached due to the VALID_COMMANDS check above
                return handleUnknownCommand(parsed.command, parsed.json);
            }
            catch (error) {
                const err = error instanceof Error ? error : new Error(String(error));
                outputError(err, { json: parsedJson, verbose: parsedVerbose });
                return 1;
            }
        },
        formatHelp() {
            return USAGE;
        },
    };
}
if (require.main === module) {
    createBabysitterCli()
        .run()
        .then((code) => {
        process.exitCode = code;
    })
        .catch((err) => {
        console.error(err);
        process.exitCode = 1;
    });
}
