// @a5c-ai/babysitter-sdk/dist/config/index.js
"use strict";
/**
 * Babysitter SDK Configuration Module
 *
 * This module provides centralized configuration management for the SDK.
 * Import from this module to access defaults, configuration helpers, and types.
 *
 * @example
 * ```ts
 * import { DEFAULTS, getConfig, validateConfig } from "@a5c/babysitter-sdk/config";
 *
 * // Use defaults directly
 * console.log(DEFAULTS.runsDir); // ".a5c/runs"
 *
 * // Get merged config with env overrides
 * const config = getConfig({ maxIterations: 500 });
 *
 * // Validate configuration
 * const result = validateConfig(config);
 * if (!result.valid) {
 *   throw new Error(result.errors.join(", "));
 * }
 * ```
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.isValidLogLevel = exports.getDefaults = exports.validateConfig = exports.getConfig = exports.CONFIG_ENV_VARS = exports.DEFAULTS = void 0;
var defaults_1 = require("./defaults");
// Constants
Object.defineProperty(exports, "DEFAULTS", { enumerable: true, get: function () { return defaults_1.DEFAULTS; } });
Object.defineProperty(exports, "CONFIG_ENV_VARS", { enumerable: true, get: function () { return defaults_1.CONFIG_ENV_VARS; } });
// Functions
Object.defineProperty(exports, "getConfig", { enumerable: true, get: function () { return defaults_1.getConfig; } });
Object.defineProperty(exports, "validateConfig", { enumerable: true, get: function () { return defaults_1.validateConfig; } });
Object.defineProperty(exports, "getDefaults", { enumerable: true, get: function () { return defaults_1.getDefaults; } });
Object.defineProperty(exports, "isValidLogLevel", { enumerable: true, get: function () { return defaults_1.isValidLogLevel; } });
