// @a5c-ai/babysitter-sdk/dist/harness/nullAdapter.js
"use strict";
/**
 * Null harness adapter (no-op fallback).
 *
 * Used when no harness is detected or the SDK is running standalone.
 * All methods return safe defaults that allow normal operation.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.createNullAdapter = createNullAdapter;
function createNullAdapter() {
    return {
        name: "none",
        isActive() {
            return false;
        },
        resolveSessionId() {
            return undefined;
        },
        resolveStateDir() {
            return undefined;
        },
        resolvePluginRoot(args) {
            return args.pluginRoot || undefined;
        },
        bindSession(opts) {
            return Promise.resolve({
                harness: "none",
                sessionId: opts.sessionId,
                error: "No harness adapter active — session not bound",
            });
        },
        handleStopHook(_args) {
            // No harness → approve (allow exit)
            process.stdout.write('{"decision":"approve"}\n');
            return Promise.resolve(0);
        },
        handleSessionStartHook(_args) {
            // No harness → nothing to do
            process.stdout.write("{}\n");
            return Promise.resolve(0);
        },
        findHookDispatcherPath() {
            return null;
        },
    };
}
