// @a5c-ai/babysitter-sdk/dist/harness/registry.js
"use strict";
/**
 * Harness adapter registry with auto-detection.
 *
 * Maintains a lazy singleton of the active adapter. On first access,
 * probes registered adapters via `isActive()` and returns the first match
 * (or the null adapter if none match).
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.detectAdapter = detectAdapter;
exports.getAdapterByName = getAdapterByName;
exports.listSupportedHarnesses = listSupportedHarnesses;
exports.getAdapter = getAdapter;
exports.setAdapter = setAdapter;
exports.resetAdapter = resetAdapter;
const claudeCode_1 = require("./claudeCode");
const nullAdapter_1 = require("./nullAdapter");
// ---------------------------------------------------------------------------
// Registry of known adapters (ordered by priority)
// ---------------------------------------------------------------------------
const knownAdapters = [
    (0, claudeCode_1.createClaudeCodeAdapter)(),
];
// ---------------------------------------------------------------------------
// Auto-detection
// ---------------------------------------------------------------------------
/**
 * Probe each registered adapter and return the first that reports active.
 * Falls back to the null adapter.
 */
function detectAdapter() {
    for (const adapter of knownAdapters) {
        if (adapter.isActive())
            return adapter;
    }
    return (0, nullAdapter_1.createNullAdapter)();
}
/**
 * Look up an adapter by harness name (e.g. "claude-code").
 * Returns null if the name is not recognized.
 */
function getAdapterByName(name) {
    for (const adapter of knownAdapters) {
        if (adapter.name === name)
            return adapter;
    }
    return null;
}
/**
 * List the names of all supported harnesses.
 */
function listSupportedHarnesses() {
    return knownAdapters.map((a) => a.name);
}
// ---------------------------------------------------------------------------
// Lazy singleton
// ---------------------------------------------------------------------------
let current = null;
/**
 * Get the active harness adapter (auto-detected on first call).
 */
function getAdapter() {
    if (!current) {
        current = detectAdapter();
    }
    return current;
}
/**
 * Override the active adapter (useful for testing).
 */
function setAdapter(adapter) {
    current = adapter;
}
/**
 * Reset the singleton so the next `getAdapter()` call re-detects.
 */
function resetAdapter() {
    current = null;
}
