// @a5c-ai/babysitter-sdk/dist/hooks/dispatcher.js
"use strict";
/**
 * Hook Dispatcher
 * Executes shell hooks from Node.js
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.findHookDispatcherPath = findHookDispatcherPath;
exports.callHook = callHook;
const node_child_process_1 = require("node:child_process");
const node_fs_1 = require("node:fs");
const path = __importStar(require("node:path"));
const defaults_1 = require("../config/defaults");
const harness_1 = require("../harness");
/**
 * Find `plugins/babysitter/hooks/hook-dispatcher.sh` by walking up from cwd.
 * This allows running from nested projects/fixtures inside a mono-repo.
 *
 * First checks the active harness adapter for a harness-specific path
 * (e.g. CLAUDE_PLUGIN_ROOT), then falls back to walking up the directory tree.
 *
 * @internal
 */
function findHookDispatcherPath(startCwd) {
    // Try harness-specific path first (e.g. CLAUDE_PLUGIN_ROOT)
    const harnessPath = (0, harness_1.getAdapter)().findHookDispatcherPath(startCwd);
    if (harnessPath)
        return harnessPath;
    let current = path.resolve(startCwd);
    // Guard against infinite loops: stop once we stop making progress.
    for (let i = 0; i < 50; i++) {
        const candidate = path.join(current, "plugins", "babysitter", "hooks", "hook-dispatcher.sh");
        if ((0, node_fs_1.existsSync)(candidate))
            return candidate;
        const parent = path.dirname(current);
        if (parent === current)
            break;
        current = parent;
    }
    return null;
}
/**
 * Call a hook by dispatching to the shell hook-dispatcher.sh
 */
async function callHook(options) {
    const { hookType, payload, cwd = process.cwd(), timeout = defaults_1.DEFAULTS.hookTimeout, throwOnFailure = false, } = options;
    const dispatcherPath = findHookDispatcherPath(cwd);
    if (!dispatcherPath) {
        return {
            hookType,
            success: false,
            error: `Hook dispatcher not found. Expected plugins/babysitter/hooks/hook-dispatcher.sh ` +
                `in ${cwd} or any parent directory.`,
            executedHooks: [],
        };
    }
    const payloadJson = JSON.stringify(payload);
    const startTime = Date.now();
    return new Promise((resolve, reject) => {
        const child = (0, node_child_process_1.spawn)("bash", [dispatcherPath, hookType], {
            cwd,
            stdio: ["pipe", "pipe", "pipe"],
            timeout,
        });
        let stdout = "";
        let stderr = "";
        let timedOut = false;
        // Send payload via stdin
        child.stdin.write(payloadJson);
        child.stdin.end();
        // Collect output
        child.stdout.on("data", (data) => {
            stdout += data.toString();
        });
        child.stderr.on("data", (data) => {
            stderr += data.toString();
        });
        // Handle timeout
        const timeoutHandle = setTimeout(() => {
            timedOut = true;
            child.kill("SIGTERM");
        }, timeout);
        child.on("error", (error) => {
            clearTimeout(timeoutHandle);
            const result = {
                hookType,
                success: false,
                error: `Failed to spawn hook dispatcher: ${error.message}`,
                executedHooks: [],
            };
            if (throwOnFailure) {
                reject(new Error(result.error));
            }
            else {
                resolve(result);
            }
        });
        child.on("close", (exitCode) => {
            clearTimeout(timeoutHandle);
            const _duration = Date.now() - startTime;
            if (timedOut) {
                const result = {
                    hookType,
                    success: false,
                    error: `Hook execution timed out after ${timeout}ms`,
                    executedHooks: [],
                };
                if (throwOnFailure) {
                    reject(new Error(result.error));
                }
                else {
                    resolve(result);
                }
                return;
            }
            // Parse execution results from stderr
            const executedHooks = parseHookExecutionSummary(stderr);
            const result = {
                hookType,
                success: exitCode === 0,
                output: stdout ? tryParseJson(stdout) : undefined,
                error: exitCode !== 0
                    ? `Hook dispatcher exited with code ${exitCode}`
                    : undefined,
                executedHooks,
            };
            if (throwOnFailure && !result.success) {
                reject(new Error(result.error ||
                    `Hook ${hookType} failed with exit code ${exitCode}`));
            }
            else {
                resolve(result);
            }
        });
    });
}
/**
 * Parse hook execution summary from stderr output
 * Looks for lines like: "plugin:logger.sh:success"
 */
function parseHookExecutionSummary(stderr) {
    const results = [];
    const lines = stderr.split("\n");
    for (const line of lines) {
        // Look for summary lines: "location:hookname:status[:exitcode]"
        const match = line.match(/^(per-repo|per-user|plugin):([^:]+):([^:]+)(?::(\d+))?$/);
        if (match) {
            const [, location, hookName, status, exitCodeStr] = match;
            results.push({
                hookPath: `unknown`, // We don't have full path in summary
                hookName,
                hookLocation: location,
                status: status,
                exitCode: exitCodeStr ? parseInt(exitCodeStr, 10) : undefined,
            });
        }
    }
    return results;
}
/**
 * Try to parse JSON, return raw string if it fails
 */
function tryParseJson(str) {
    try {
        return JSON.parse(str);
    }
    catch {
        return str.trim();
    }
}
