// @a5c-ai/babysitter-sdk/dist/storage/createRunDir.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createRunDir = createRunDir;
const fs_1 = require("fs");
const path_1 = __importDefault(require("path"));
const paths_1 = require("./paths");
const atomic_1 = require("./atomic");
const clock_1 = require("./clock");
const GITIGNORE_CONTENT = `state/\ntasks/*/artifacts/\nblobs/\norphaned/\n`;
async function createRunDir(options) {
    const runDir = (0, paths_1.getRunDir)(options.runsRoot, options.runId);
    await fs_1.promises.mkdir(runDir, { recursive: true });
    await Promise.all([
        fs_1.promises.mkdir((0, paths_1.getJournalDir)(runDir), { recursive: true }),
        fs_1.promises.mkdir((0, paths_1.getTasksDir)(runDir), { recursive: true }),
        fs_1.promises.mkdir((0, paths_1.getBlobsDir)(runDir), { recursive: true }),
        fs_1.promises.mkdir((0, paths_1.getStateDir)(runDir), { recursive: true }),
        fs_1.promises.mkdir(path_1.default.join(runDir, paths_1.ORPHANED_DIR), { recursive: true }),
        fs_1.promises.mkdir(path_1.default.join(runDir, paths_1.PROCESS_DIR), { recursive: true }),
    ]);
    await (0, atomic_1.writeFileAtomic)(path_1.default.join(runDir, ".gitignore"), GITIGNORE_CONTENT);
    const layoutVersion = options.layoutVersion ?? paths_1.DEFAULT_LAYOUT_VERSION;
    const entrypoint = resolveEntrypoint(options);
    const createdAt = (0, clock_1.getClockIsoString)();
    const metadata = {
        runId: options.runId,
        request: options.request,
        processId: options.processId ?? options.request ?? options.runId,
        entrypoint,
        processPath: entrypoint.importPath,
        processRevision: options.processRevision,
        layoutVersion,
        createdAt,
        ...(options.prompt !== undefined ? { prompt: options.prompt } : {}),
    };
    if (options.extraMetadata) {
        Object.assign(metadata, options.extraMetadata);
    }
    await (0, atomic_1.writeFileAtomic)(path_1.default.join(runDir, paths_1.RUN_METADATA_FILE), JSON.stringify(metadata, null, 2) + "\n");
    if (options.inputs !== undefined) {
        await (0, atomic_1.writeFileAtomic)(path_1.default.join(runDir, paths_1.INPUTS_FILE), JSON.stringify(options.inputs, null, 2) + "\n");
    }
    return { runDir, metadata };
}
function resolveEntrypoint(options) {
    if (options.entrypoint?.importPath) {
        return {
            importPath: options.entrypoint.importPath,
            exportName: options.entrypoint.exportName ?? "process",
        };
    }
    const importPath = options.processPath ?? "./process.js";
    return {
        importPath,
        exportName: "process",
    };
}
