// @a5c-ai/babysitter-sdk/dist/storage/paths.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_LAYOUT_VERSION = exports.STATE_FILE = exports.LOCK_FILE = exports.INPUTS_FILE = exports.RUN_METADATA_FILE = exports.PROCESS_DIR = exports.ORPHANED_DIR = exports.STATE_DIR = exports.BLOBS_DIR = exports.TASKS_DIR = exports.JOURNAL_DIR = void 0;
exports.getRunDir = getRunDir;
exports.getJournalDir = getJournalDir;
exports.getTasksDir = getTasksDir;
exports.getBlobsDir = getBlobsDir;
exports.getStateDir = getStateDir;
exports.getStateFile = getStateFile;
exports.getLockPath = getLockPath;
const path_1 = __importDefault(require("path"));
const defaults_1 = require("../config/defaults");
exports.JOURNAL_DIR = "journal";
exports.TASKS_DIR = "tasks";
exports.BLOBS_DIR = "blobs";
exports.STATE_DIR = "state";
exports.ORPHANED_DIR = "orphaned";
exports.PROCESS_DIR = "process";
exports.RUN_METADATA_FILE = "run.json";
exports.INPUTS_FILE = "inputs.json";
exports.LOCK_FILE = "run.lock";
exports.STATE_FILE = "state.json";
/**
 * Default storage layout version.
 * @see DEFAULTS.layoutVersion for the centralized default
 */
exports.DEFAULT_LAYOUT_VERSION = defaults_1.DEFAULTS.layoutVersion;
function getRunDir(runsRoot, runId) {
    return path_1.default.join(runsRoot, runId);
}
function getJournalDir(runDir) {
    return path_1.default.join(runDir, exports.JOURNAL_DIR);
}
function getTasksDir(runDir) {
    return path_1.default.join(runDir, exports.TASKS_DIR);
}
function getBlobsDir(runDir) {
    return path_1.default.join(runDir, exports.BLOBS_DIR);
}
function getStateDir(runDir) {
    return path_1.default.join(runDir, exports.STATE_DIR);
}
function getStateFile(runDir) {
    return path_1.default.join(getStateDir(runDir), exports.STATE_FILE);
}
function getLockPath(runDir) {
    return path_1.default.join(runDir, exports.LOCK_FILE);
}
