// @a5c-ai/babysitter-sdk/dist/cli/commands/runExecuteTasks.js
"use strict";
/**
 * run:execute-tasks command - Execute pending node tasks in a run
 *
 * This command replaces the shell-based task execution logic in native-orchestrator.sh.
 * It:
 * 1. Loads the journal and builds the effect index
 * 2. Filters pending effects by kind (default "node")
 * 3. Slices to maxTasks (default 3)
 * 4. For each task: reads task.json, spawns `node`, captures stdout/stderr, commits result
 * 5. Returns a summary of executed tasks
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.runExecuteTasks = runExecuteTasks;
const node_child_process_1 = require("node:child_process");
const node_fs_1 = require("node:fs");
const path = __importStar(require("node:path"));
const effectIndex_1 = require("../../runtime/replay/effectIndex");
const commitEffectResult_1 = require("../../runtime/commitEffectResult");
const tasks_1 = require("../../storage/tasks");
const defaults_1 = require("../../config/defaults");
function log(verbose, ...args) {
    if (verbose) {
        console.error(`[run:execute-tasks]`, ...args);
    }
}
/**
 * Resolve a path relative to runDir unless it is already absolute.
 */
function resolveRelative(runDir, ref) {
    if (path.isAbsolute(ref) || /^[A-Za-z]:[\\/]/.test(ref)) {
        return ref;
    }
    return path.join(runDir, ref);
}
/**
 * Make a path relative to runDir, using forward slashes.
 */
function toRunRelative(runDir, absolute) {
    return path.relative(runDir, absolute).replace(/\\/g, "/");
}
/**
 * Spawn a node process for the given task and capture its output.
 */
function spawnNodeTask(options) {
    return new Promise((resolve) => {
        const child = (0, node_child_process_1.spawn)("node", [options.entry, ...options.args], {
            cwd: options.cwd,
            env: { ...process.env, ...options.env },
            stdio: ["ignore", "pipe", "pipe"],
            timeout: options.timeout,
        });
        const stdoutChunks = [];
        const stderrChunks = [];
        child.stdout.on("data", (chunk) => stdoutChunks.push(chunk));
        child.stderr.on("data", (chunk) => stderrChunks.push(chunk));
        child.on("close", (code) => {
            resolve({
                exitCode: code ?? 1,
                stdout: Buffer.concat(stdoutChunks).toString("utf8"),
                stderr: Buffer.concat(stderrChunks).toString("utf8"),
            });
        });
        child.on("error", (err) => {
            resolve({
                exitCode: 1,
                stdout: Buffer.concat(stdoutChunks).toString("utf8"),
                stderr: `${Buffer.concat(stderrChunks).toString("utf8")}\n${err.message}`,
            });
        });
    });
}
/**
 * Execute a single task: read its definition, spawn node, commit the result.
 */
async function executeOneTask(runDir, record, timeout, verbose) {
    const effectId = record.effectId;
    const label = record.label ?? null;
    log(verbose, `Executing: ${effectId}${label ? ` (${label})` : ""}`);
    // Read task.json
    const taskDef = await (0, tasks_1.readTaskDefinition)(runDir, effectId);
    if (!taskDef) {
        log(verbose, `Missing task definition for ${effectId}`);
        const errorPayload = {
            name: "Error",
            message: `Missing task definition for effect ${effectId}`,
        };
        await (0, commitEffectResult_1.commitEffectResult)({
            runDir,
            effectId,
            invocationKey: record.invocationKey,
            result: {
                status: "error",
                error: errorPayload,
                startedAt: new Date().toISOString(),
                finishedAt: new Date().toISOString(),
            },
        });
        return {
            effectId,
            label,
            status: "error",
            exitCode: 1,
            durationMs: 0,
            stdoutRef: null,
            stderrRef: null,
            resultRef: null,
            error: errorPayload,
        };
    }
    // Extract node config
    const nodeConfig = (taskDef.node ?? {});
    const entry = nodeConfig.entry;
    if (!entry) {
        log(verbose, `Missing node.entry in task.json for ${effectId}`);
        const errorPayload = {
            name: "Error",
            message: `Missing node.entry in task definition for effect ${effectId}`,
        };
        await (0, commitEffectResult_1.commitEffectResult)({
            runDir,
            effectId,
            invocationKey: record.invocationKey,
            result: {
                status: "error",
                error: errorPayload,
                startedAt: new Date().toISOString(),
                finishedAt: new Date().toISOString(),
            },
        });
        return {
            effectId,
            label,
            status: "error",
            exitCode: 1,
            durationMs: 0,
            stdoutRef: null,
            stderrRef: null,
            resultRef: null,
            error: errorPayload,
        };
    }
    const nodeArgs = nodeConfig.args ?? [];
    const taskTimeout = nodeConfig.timeoutMs ?? timeout;
    // Extract IO config
    const ioConfig = (taskDef.io ?? {});
    const inputRef = ioConfig.inputJsonPath ?? `tasks/${effectId}/inputs.json`;
    const outputRef = ioConfig.outputJsonPath ?? `tasks/${effectId}/result.json`;
    const stdoutRef = ioConfig.stdoutPath ?? `tasks/${effectId}/stdout.log`;
    const stderrRef = ioConfig.stderrPath ?? `tasks/${effectId}/stderr.log`;
    // Resolve paths relative to runDir
    const inputAbs = resolveRelative(runDir, inputRef);
    const outputAbs = resolveRelative(runDir, outputRef);
    const stdoutAbs = resolveRelative(runDir, stdoutRef);
    const stderrAbs = resolveRelative(runDir, stderrRef);
    const entryAbs = resolveRelative(runDir, entry);
    const cwdAbs = nodeConfig.cwd ? resolveRelative(runDir, nodeConfig.cwd) : runDir;
    // Ensure directories exist
    await Promise.all([
        node_fs_1.promises.mkdir(path.dirname(inputAbs), { recursive: true }),
        node_fs_1.promises.mkdir(path.dirname(outputAbs), { recursive: true }),
        node_fs_1.promises.mkdir(path.dirname(stdoutAbs), { recursive: true }),
        node_fs_1.promises.mkdir(path.dirname(stderrAbs), { recursive: true }),
    ]);
    // Stage inputs.json
    const inputsRefField = taskDef.inputsRef;
    if (inputsRefField) {
        const sourceAbs = resolveRelative(runDir, inputsRefField);
        try {
            await node_fs_1.promises.copyFile(sourceAbs, inputAbs);
        }
        catch {
            // If the source doesn't exist, write an empty object
            await node_fs_1.promises.writeFile(inputAbs, "{}\n", "utf8");
        }
    }
    else {
        // Extract inline inputs or default to {}
        const inputs = taskDef.inputs ?? {};
        await node_fs_1.promises.writeFile(inputAbs, JSON.stringify(inputs) + "\n", "utf8");
    }
    // Build environment for the child process
    const taskEnv = {
        ...(nodeConfig.env ?? {}),
        BABYSITTER_INPUT_JSON: inputAbs,
        BABYSITTER_OUTPUT_JSON: outputAbs,
        BABYSITTER_STDOUT_PATH: stdoutAbs,
        BABYSITTER_STDERR_PATH: stderrAbs,
        BABYSITTER_EFFECT_ID: effectId,
    };
    // Spawn the node process
    const startedAt = new Date().toISOString();
    const startMs = Date.now();
    const spawnResult = await spawnNodeTask({
        entry: entryAbs,
        args: nodeArgs,
        cwd: cwdAbs,
        env: taskEnv,
        stdoutPath: stdoutAbs,
        stderrPath: stderrAbs,
        timeout: taskTimeout,
    });
    const finishedAt = new Date().toISOString();
    const durationMs = Date.now() - startMs;
    // Write stdout and stderr to disk
    await node_fs_1.promises.writeFile(stdoutAbs, spawnResult.stdout, "utf8");
    await node_fs_1.promises.writeFile(stderrAbs, spawnResult.stderr, "utf8");
    const stdoutRelRef = toRunRelative(runDir, stdoutAbs);
    const stderrRelRef = toRunRelative(runDir, stderrAbs);
    // Commit the result
    if (spawnResult.exitCode === 0) {
        // Read the output file produced by the task (if any)
        let value = undefined;
        try {
            const outputContents = await node_fs_1.promises.readFile(outputAbs, "utf8");
            const trimmed = outputContents.trim();
            if (trimmed.length > 0) {
                value = JSON.parse(trimmed);
            }
        }
        catch {
            // Output file may not exist; that is acceptable
        }
        const committed = await (0, commitEffectResult_1.commitEffectResult)({
            runDir,
            effectId,
            invocationKey: record.invocationKey,
            result: {
                status: "ok",
                value,
                stdoutRef: stdoutRelRef,
                stderrRef: stderrRelRef,
                startedAt,
                finishedAt,
            },
        });
        log(verbose, `Posted result: ${effectId} (ok, ${durationMs}ms)`);
        return {
            effectId,
            label,
            status: "ok",
            exitCode: spawnResult.exitCode,
            durationMs,
            stdoutRef: stdoutRelRef,
            stderrRef: stderrRelRef,
            resultRef: committed.resultRef ?? null,
        };
    }
    else {
        const errorPayload = {
            name: "Error",
            message: `Node task exited non-zero (exitCode=${spawnResult.exitCode})`,
        };
        const committed = await (0, commitEffectResult_1.commitEffectResult)({
            runDir,
            effectId,
            invocationKey: record.invocationKey,
            result: {
                status: "error",
                error: errorPayload,
                stdoutRef: stdoutRelRef,
                stderrRef: stderrRelRef,
                startedAt,
                finishedAt,
            },
        });
        log(verbose, `Posted error: ${effectId} (exitCode=${spawnResult.exitCode}, ${durationMs}ms)`);
        return {
            effectId,
            label,
            status: "error",
            exitCode: spawnResult.exitCode,
            durationMs,
            stdoutRef: stdoutRelRef,
            stderrRef: stderrRelRef,
            resultRef: committed.resultRef ?? null,
            error: errorPayload,
        };
    }
}
async function runExecuteTasks(options) {
    const { runDir, verbose, dryRun, } = options;
    const maxTasks = options.maxTasks ?? 3;
    const kind = options.kind ?? "node";
    const config = (0, defaults_1.getConfig)();
    const timeout = options.timeout ?? config.nodeTaskTimeout;
    log(verbose, `Scanning for pending ${kind} tasks in ${runDir} (maxTasks=${maxTasks})`);
    // Build the effect index from journal
    const effectIndex = await (0, effectIndex_1.buildEffectIndex)({ runDir });
    const pendingEffects = effectIndex.listPendingEffects();
    if (pendingEffects.length === 0) {
        log(verbose, "No pending effects found");
        return {
            action: "none",
            count: 0,
            reason: "no-pending-effects",
            tasks: [],
        };
    }
    // Filter by kind
    const matchingEffects = pendingEffects.filter((record) => record.kind?.toLowerCase() === kind.toLowerCase());
    if (matchingEffects.length === 0) {
        log(verbose, `No pending effects matching kind="${kind}"`);
        return {
            action: "none",
            count: 0,
            reason: "no-matching-tasks",
            tasks: [],
        };
    }
    // Slice to maxTasks
    const tasksToExecute = matchingEffects.slice(0, maxTasks);
    log(verbose, `Found ${matchingEffects.length} matching tasks, executing ${tasksToExecute.length}`);
    if (dryRun) {
        const dryRunTasks = tasksToExecute.map((record) => ({
            effectId: record.effectId,
            label: record.label ?? null,
            status: "ok",
            exitCode: 0,
            durationMs: 0,
            stdoutRef: null,
            stderrRef: null,
            resultRef: null,
        }));
        return {
            action: "none",
            count: tasksToExecute.length,
            reason: "dry-run",
            tasks: dryRunTasks,
        };
    }
    // Execute tasks sequentially
    const taskSummaries = [];
    for (const record of tasksToExecute) {
        const summary = await executeOneTask(runDir, record, timeout, verbose);
        taskSummaries.push(summary);
    }
    return {
        action: "executed-tasks",
        count: taskSummaries.length,
        reason: "auto-runnable-tasks",
        tasks: taskSummaries,
    };
}
